{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download and explore GTEx datasets.\n",
    "Namely, the eQTL datasets (one file per tissue)  and the gene median TPM dataset.  \n",
    "Data Source link: https://gtexportal.org/home/datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## As of March 22 2023 I am using [polars](https://www.pola.rs) library to process the larger eQTL (70 million) dataset. In order to load all 50 tissue files (each containing eQTLs) with polars, the files must be unzipped (actually gunzipped) which was done with the linux command ` gunzip *.txt.gz`\n",
    "### The median TPM workflow (second half of this notebook) still uses pandas which is fine.\n",
    "### HOWEVER, Updates to the GTEX_EXP (median TPM) workflow are going to be made in THIS notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The end of this workflow is different from the original GTEx.ipynb notebook located in /Users/stearb/Dropbox/CHOP/R03/code/GTEx, we are using Jonathan Silversteins workflow for the Neo4j CSV creation (meaning the files produced by this workflow will be the inputs into JS's workflow) ...so we only need to create 2 files, a nodes.tsv and an edges.tsv (instead of the ~6 files, CUIs, CUI-CUIs, Code-CUIs, Terms, etc.) The nodes aand edges files are then inputs into the OWLNETS python script which transforms them into the 11 CSV files that get appended to the base UMLS CSVs.\n",
    "\n",
    "#### The guide for how to create these new nodes and edges files can be found in the Data Distillerys [github](https://github.com/dbmi-pitt/UBKG/tree/main/user%20guide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## As of Feb 18 2023, the new format for GTEX_EQTLs is: \n",
    "\n",
    "#### `GTEX_EQTL eQTL_chr{CHROMOSOME_NUMBER}_{LOCATION}_{REF}_{ALT}_b38_{TISSUE}`\n",
    "\n",
    "#### ie) `GTEX_EQTL eQTL_chr11_538371_G_A_b38_Skin_Not_Sun_Exposed_Suprapubic`\n",
    "\n",
    "(changed from the old format of rsID-HGNC_ID-tissue)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## As of March 7th we are using the `signif_variant_gene_pairs.txt` files and NOT the `egenes.txt` files. `signif_variant_gene_pairs.txt.gz` contains ~71 million eQTLs while `egenes` contains only ~2.1 million"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## As of May 2023 the SABs for the GTEx eQTL and GTEx expression datasets will be GTEXEQTL and GTEXEXP, respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!jupyter nbconvert --to script GTEx_JS.ipynb\n",
    "#!sed -i '' '/.head(/d' GTEx_JS.ipynb\n",
    "#!sed -i '' '/^#/d' GTEx_JS.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "from polars import Config\n",
    "Config.set_fmt_str_lengths = 1000\n",
    "import numpy as np\n",
    "import os\n",
    "from collections import Counter\n",
    "import pyarrow\n",
    "from cmapPy.pandasGEXpress.parse_gct import parse\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modeling Notes/Questions \n",
    "Median Expression \n",
    "* Median Expression Code nodes have Interval nodes as Terms. Each Interval Term node is connected to many Median Expression Code Nodes, is this allowed? (Code nodes sharing Term nodes)\n",
    "* Median Expression Code nodes CodeID and CODE (see abover)\n",
    "\n",
    "eQTL\n",
    "* Currently using HGNC/UBERON/dbSNP Concept nodes to uniquely identify every eQTL \n",
    "* Instead,use the variant_id to link the eqtl node (with variant_id) to UBERON/HGNC nodes\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Concepts can link to multiple Codes from different ontologies\n",
    "    * Ex) CUI C1416553 links to HGNC and OMIM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  First, Get the Sample Annotations, which contain sample and uberon to tissue mappings (GTEx--UBERON mappings).\n",
    "Data Source: https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22951, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>SMUBRID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SMTS        SMTSD  SMUBRID\n",
       "0  Blood  Whole Blood  0013756\n",
       "1  Blood  Whole Blood  0013756\n",
       "2  Blood  Whole Blood  0013756"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#!curl --insecure https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt > gtex_sample_annotations.txt\n",
    "samp_annos = pd.read_csv('gtex_sample_annotations_select.txt',sep='\\t')\n",
    "print(samp_annos.shape)\n",
    "samp_annos.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SAMPID_2_TISSUE.to_csv('gtex_sample_annotations_select.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22951, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>SMUBRID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brain</td>\n",
       "      <td>Brain Frontal Cortex BA9</td>\n",
       "      <td>0009834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brain</td>\n",
       "      <td>Brain Frontal Cortex BA9</td>\n",
       "      <td>0009834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SMTS                     SMTSD  SMUBRID\n",
       "0  Blood               Whole Blood  0013756\n",
       "1  Blood               Whole Blood  0013756\n",
       "2  Blood               Whole Blood  0013756\n",
       "3  Brain  Brain Frontal Cortex BA9  0009834\n",
       "4  Brain  Brain Frontal Cortex BA9  0009834"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select the 3 columns  we need from the annotation dataset     # dont actually need sample id\n",
    "# SAMPID = SAMPLE ID\n",
    "# SMTS = Tissue\n",
    "# SMTSD = More specific Tissue\n",
    "# SMUBRID = UBERON ID\n",
    "SAMPID_2_TISSUE = samp_annos[['SMTS','SMTSD','SMUBRID']] # 'SAMPID',\n",
    "\n",
    "# Remove dashes, -, and parentheses so the tissue strings match the tissue strings from the eqtl datasets.\n",
    "SAMPID_2_TISSUE['SMTSD'] = SAMPID_2_TISSUE['SMTSD'].str.replace(' - ',' ').str.replace('\\(','').str.replace('\\)','')\n",
    "\n",
    "# Get only unique tissue IDs for mapping tissues--UBERON codes later.\n",
    "SAMPID_2_TISSUE_unique = SAMPID_2_TISSUE.drop_duplicates('SMTSD')  # .drop('SAMPID',axis=1)\n",
    "SAMPID_2_TISSUE_unique.rename(columns={'SMTSD':'tissue'},inplace=True)\n",
    "\n",
    "print(SAMPID_2_TISSUE.shape)\n",
    "SAMPID_2_TISSUE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  Also, get the master list of HGNC IDs (that maps gene names to HGNC IDs)\n",
    "Data Source: ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/non_alt_loci_set.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!curl ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/non_alt_loci_set.txt > hgnc_master.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(43653, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      hgnc_id    symbol  ensembl_gene_id\n",
       "0      HGNC:5      A1BG  ENSG00000121410\n",
       "1  HGNC:37133  A1BG-AS1  ENSG00000268895\n",
       "2  HGNC:24086      A1CF  ENSG00000148584"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#!curl ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/non_alt_loci_set.txt > hgnc_master.txt\n",
    "hgnc_master = pd.read_csv('hgnc_master.txt',sep='\\t')\n",
    "hgnc_select = hgnc_master[['hgnc_id','symbol','ensembl_gene_id']]\n",
    "print(hgnc_select.shape)\n",
    "hgnc_select.head(3)\n",
    "#hgnc_master[['hgnc_id','symbol']].to_csv('hgnc_master_2cols.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import eQTL data (as of May 20th, 2023 we are filtering eQTLs to reduce the size of the resulting nodes/edges files. We are only including 'common' eQTLs, ones that are present in every one of the 49 tissues)\n",
    "File:  GTEx_Analysis_v8_eQTL/  \n",
    "(Contains 49 tissue files)  \n",
    "\n",
    "[Nominal P-value Explanation](https://stats.stackexchange.com/questions/536116/what-is-the-concepts-of-nominal-and-actual-significance-level)\n",
    " \n",
    "[Ensembl Gene ID Period Explanation](https://useast.ensembl.org/Help/Faq?id=488)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Not using this for loop to load in eqtls anymore (that code is at the bottom of this nb). Use the next block to load in the single 71m eqtl dataset using polars "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.97 s, sys: 2.4 s, total: 6.37 s\n",
      "Wall time: 4.16 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "eqtl = pl.read_parquet('/Users/stearb/Desktop/DESKTOP_TRANSFER/R03_local/data/gtex/eqtls_all_71m.parquet')\n",
    "eqtl = eqtl.drop('__index_level_0__')\n",
    "#eqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hgnc_select# = hgnc_select.rename(columns={'ensembl_gene_id':'gene_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eqtl = eqtl.rename({'gene_id':'ensembl_gene_id'})\n",
    "#eqtl['ensembl_gene_id'] = eqtl['gene_id'].apply(lambda x: x.split('.')[0]) \n",
    "#eqtl.select(pl.col('gene_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11 s, sys: 29.7 s, total: 40.7 s\n",
      "Wall time: 29.7 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(71589767, 6)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "###### leave the '.' in the 'gene_id' column (the ensembl code uses a .N to specify the \n",
    "# Nth revision of a genes definition). The '.' is ok to have in the code now that we are ingesting ensembl ids\n",
    "# the HGNC ensembl genes dont have the '.' though so we can still make this col, just to merge on and then\n",
    "# delete it.\n",
    "eqtl = eqtl.with_columns(pl.col(\"gene_id\").str.split(\".\").arr.get(0).alias(\"ensembl_gene_id\"))\n",
    "df = eqtl.join(pl.from_pandas(hgnc_select),on='ensembl_gene_id',how='left').drop('ensembl_gene_id')\n",
    "del eqtl\n",
    "df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## If we use ensembl ids we dont need to drop nulls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop_nulls() # drops 71m to ~61m   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get heart eqtls\n",
    "#df.select(pl.col('tissue')).unique().to_pandas()\n",
    "#heart_tissues = ['Artery Coronary','Heart Atrial Appendage','Artery Aorta','Heart Left Ventricle']\n",
    "#df_heart = df.filter(pl.col('tissue').is_in(heart_tissues))\n",
    "#df_heart = df_heart\n",
    "#df_heart\n",
    "#df_heart.write_parquet('gtex_heart_eqtls.parquet')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specify delimiter for GTEX_EQTL CODEs. It used to be '_', now we are using '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "EQTL_CODE_DELIM = '.'\n",
    "SAB = 'GTEXEQTL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.with_columns(pl.col('tissue').str.replace_all(' ',EQTL_CODE_DELIM).alias('tissue_underscore'))\\\n",
    "       .with_columns(pl.col('variant_id').str.replace_all('_',EQTL_CODE_DELIM).alias('variant_id'))\n",
    "\n",
    "df = df.with_columns(pl.map([pl.col('variant_id'),pl.col('tissue_underscore')], \n",
    "                      lambda s: f\"{SAB} eQTL{EQTL_CODE_DELIM}\" \n",
    "                              +s[0] +EQTL_CODE_DELIM+ s[1]).alias(\"CodeID_gtex\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.select('CodeID_gtex').head(3).to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.52 s, sys: 3.65 s, total: 6.17 s\n",
      "Wall time: 1.92 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# MERGE IN UBERON CODEs\n",
    "eqtl_ub = df.join(pl.from_pandas(SAMPID_2_TISSUE_unique),on='tissue',how='left')\n",
    "eqtl_ub = eqtl_ub.rename({'SMUBRID':'UBERON_code'}) # 'gene_name': 'symbol',\n",
    "eqtl_all_GTEx = eqtl_ub\n",
    "del eqtl_ub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optionally filter by eqtl frequency -- do this first and then do all the other merges\n",
    "### Join the variant_id field with the gene_id so we can see how many times this new string appears. should be once in every tissue so we can just filter by the frequency of this new string\n",
    "If the eqtl isnt in at least 5 tissues, we can drop it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTEx = eqtl_all_GTEx.with_columns(pl.map([pl.col('variant_id'),pl.col('gene_id')], \n",
    "                      lambda s: s[0] +EQTL_CODE_DELIM+ s[1]).alias(\"eqtl_gene\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get frequency counts of this new string\n",
    "cnts = eqtl_all_GTEx.select([pl.col(\"eqtl_gene\").value_counts()])\n",
    "# split the array into 2 cols\n",
    "cnts = cnts.with_columns(pl.col('eqtl_gene').struct.field(\"counts\").alias(\"eqtl_gene_freq\"))\n",
    "cnts = cnts.with_columns(pl.col('eqtl_gene').struct.field(\"eqtl_gene\").alias(\"eqtl_gene_str\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ONLY INCLUDE EQTL IF ITS IN 49 (all of them) TISSUES\n",
    "#### ~71m  `-->`  ~2m eQTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 49 tissues in total\n",
    "#eqtl_all_GTEx.select(pl.col('tissue').unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42086,)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TISSUE_FREQUENCY = 49\n",
    "filtered_ids = cnts.filter(pl.col('eqtl_gene_freq') >= TISSUE_FREQUENCY).select(pl.col('eqtl_gene_str')).to_series()\n",
    "filtered_ids.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2089174, 11)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_all_GTEx = eqtl_all_GTEx.filter(pl.col('eqtl_gene').is_in(filtered_ids))\n",
    "eqtl_all_GTEx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Select rare eqtls -- NOT doing this\n",
    "#TISSUE_FREQUENCY = 1\n",
    "#filtered_ids = cnts.filter(pl.col('eqtl_gene_freq') <= TISSUE_FREQUENCY).select(pl.col('eqtl_gene_str')).to_series()\n",
    "#filtered_ids.shape\n",
    "#eqtl_all_GTEx_filt = eqtl_all_GTEx.filter(pl.col('eqtl_gene').is_in(filtered_ids))\n",
    "#eqtl_all_GTEx_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2089174, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>gene_id</th><th>pval_beta</th><th>variant_id</th><th>tissue</th><th>hgnc_id</th><th>symbol</th><th>tissue_underscore</th><th>CodeID_gtex</th><th>SMTS</th><th>UBERON_code</th><th>eqtl_gene</th></tr><tr><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1434243.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1434243.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1497758.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1497758.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499000.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499000.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499128.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499128.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499639.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499639.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1500526.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1500526.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1520463.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1520463.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1531013.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1531013.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1534481.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1534481.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1558462.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1558462.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1562701.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1562701.C…</td></tr><tr><td>&quot;ENSG0000018933…</td><td>5.6663e-115</td><td>&quot;chr1.1638879.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:33941&quot;</td><td>&quot;SLC35E2B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1638879.C…</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119551999…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119551999…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119552339…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119552339…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119552594…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119552594…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119553799…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119553799…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119555676…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119555676…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119557564…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119557564…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119558129…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119558129…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119563666…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119563666…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119567350…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119567350…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119569731…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119569731…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119570829…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119570829…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119571613…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119571613…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2_089_174, 11)\n",
       "┌────────────┬────────────┬──────────┬───────────┬───┬───────────┬─────────┬───────────┬───────────┐\n",
       "│ gene_id    ┆ pval_beta  ┆ variant_ ┆ tissue    ┆ … ┆ CodeID_gt ┆ SMTS    ┆ UBERON_co ┆ eqtl_gene │\n",
       "│ ---        ┆ ---        ┆ id       ┆ ---       ┆   ┆ ex        ┆ ---     ┆ de        ┆ ---       │\n",
       "│ str        ┆ f64        ┆ ---      ┆ str       ┆   ┆ ---       ┆ str     ┆ ---       ┆ str       │\n",
       "│            ┆            ┆ str      ┆           ┆   ┆ str       ┆         ┆ str       ┆           │\n",
       "╞════════════╪════════════╪══════════╪═══════════╪═══╪═══════════╪═════════╪═══════════╪═══════════╡\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.143 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1434 │\n",
       "│ 60072.19   ┆            ┆ 4243.G.A ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 243.G.A.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1434243. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ G.A.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1497 │\n",
       "│ 60072.19   ┆            ┆ 7758.C.T ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 758.C.T.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1497758. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.T.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1499 │\n",
       "│ 60072.19   ┆            ┆ 9000.C.A ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 000.C.A.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1499000. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.A.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1499 │\n",
       "│ 60072.19   ┆            ┆ 9128.C.T ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 128.C.T.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1499128. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.T.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ …          ┆ …          ┆ …        ┆ …         ┆ … ┆ …         ┆ …       ┆ …         ┆ …         │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 567350.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 67350.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11956735 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 0.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 569731.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 69731.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11956973 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 1.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 570829.C ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 70829.C.T │\n",
       "│            ┆            ┆ .T.b38   ┆           ┆   ┆ .11957082 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 9.C.T…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 571613.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 71613.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11957161 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 3.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "└────────────┴────────────┴──────────┴───────────┴───┴───────────┴─────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_all_GTEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eqtl_all_GTEx.select('CodeID_gtex').to_pandas().to_csv('/Users/stearb/Desktop/common_eqtls.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "def fill_missing_cols(df):\n",
    "    \n",
    "    if 'node_id' not in df.columns:\n",
    "        raise ValueError('Must have at least a \"node_id\" column.')\n",
    "    \n",
    "    all_cols = set([ 'node_label', 'node_synonyms', 'node_dbxrefs',\n",
    "            'node_definition','node_namespace','value','lowerbound','upperbound','unit'])\n",
    "    \n",
    "    missing_cols = list(all_cols - set(df.columns))\n",
    "    \n",
    "    nan_cols_df = pd.DataFrame(np.full([len(df), len(missing_cols)], np.nan),columns=missing_cols)\n",
    "\n",
    "    if isinstance(df, pd.DataFrame):\n",
    "        nan_cols_df.index = df.index\n",
    "        return pd.concat([df,nan_cols_df],axis=1)\n",
    "    \n",
    "    elif isinstance(df, pl.DataFrame):\n",
    "        # no index for polars\n",
    "        return pl.concat([df,pl.from_pandas(nan_cols_df)],how='horizontal')\n",
    "    else:\n",
    "        raise ValueError(f'Must Pass either a pandas DataFrame or a polars DataFrame but recieved \"{type(df)}\".')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2089174, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>gene_id</th><th>pval_beta</th><th>variant_id</th><th>tissue</th><th>hgnc_id</th><th>symbol</th><th>tissue_underscore</th><th>CodeID_gtex</th><th>SMTS</th><th>UBERON_code</th><th>eqtl_gene</th></tr><tr><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1434243.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1434243.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1497758.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1497758.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499000.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499000.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499128.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499128.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1499639.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1499639.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1500526.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1500526.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1520463.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1520463.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1531013.G…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1531013.G…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1534481.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1534481.C…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1558462.T…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1558462.T…</td></tr><tr><td>&quot;ENSG0000016007…</td><td>5.6862e-41</td><td>&quot;chr1.1562701.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:24007&quot;</td><td>&quot;ATAD3B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1562701.C…</td></tr><tr><td>&quot;ENSG0000018933…</td><td>5.6663e-115</td><td>&quot;chr1.1638879.C…</td><td>&quot;Cells Cultured…</td><td>&quot;HGNC:33941&quot;</td><td>&quot;SLC35E2B&quot;</td><td>&quot;Cells.Cultured…</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Skin&quot;</td><td>&quot;EFO_0002009&quot;</td><td>&quot;chr1.1638879.C…</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119551999…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119551999…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119552339…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119552339…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119552594…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119552594…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119553799…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119553799…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119555676…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119555676…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119557564…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119557564…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119558129…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119558129…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119563666…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119563666…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119567350…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119567350…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119569731…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119569731…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119570829…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119570829…</td></tr><tr><td>&quot;ENSG0000001861…</td><td>4.3164e-24</td><td>&quot;chrX.119571613…</td><td>&quot;Stomach&quot;</td><td>&quot;HGNC:26239&quot;</td><td>&quot;STEEP1&quot;</td><td>&quot;Stomach&quot;</td><td>&quot;GTEXEQTL eQTL.…</td><td>&quot;Stomach&quot;</td><td>&quot;0000945&quot;</td><td>&quot;chrX.119571613…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2_089_174, 11)\n",
       "┌────────────┬────────────┬──────────┬───────────┬───┬───────────┬─────────┬───────────┬───────────┐\n",
       "│ gene_id    ┆ pval_beta  ┆ variant_ ┆ tissue    ┆ … ┆ CodeID_gt ┆ SMTS    ┆ UBERON_co ┆ eqtl_gene │\n",
       "│ ---        ┆ ---        ┆ id       ┆ ---       ┆   ┆ ex        ┆ ---     ┆ de        ┆ ---       │\n",
       "│ str        ┆ f64        ┆ ---      ┆ str       ┆   ┆ ---       ┆ str     ┆ ---       ┆ str       │\n",
       "│            ┆            ┆ str      ┆           ┆   ┆ str       ┆         ┆ str       ┆           │\n",
       "╞════════════╪════════════╪══════════╪═══════════╪═══╪═══════════╪═════════╪═══════════╪═══════════╡\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.143 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1434 │\n",
       "│ 60072.19   ┆            ┆ 4243.G.A ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 243.G.A.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1434243. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ G.A.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1497 │\n",
       "│ 60072.19   ┆            ┆ 7758.C.T ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 758.C.T.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1497758. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.T.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1499 │\n",
       "│ 60072.19   ┆            ┆ 9000.C.A ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 000.C.A.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1499000. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.A.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ ENSG000001 ┆ 5.6862e-41 ┆ chr1.149 ┆ Cells     ┆ … ┆ GTEXEQTL  ┆ Skin    ┆ EFO_00020 ┆ chr1.1499 │\n",
       "│ 60072.19   ┆            ┆ 9128.C.T ┆ Cultured  ┆   ┆ eQTL.chr1 ┆         ┆ 09        ┆ 128.C.T.b │\n",
       "│            ┆            ┆ .b38     ┆ fibroblas ┆   ┆ .1499128. ┆         ┆           ┆ 38.ENSG00 │\n",
       "│            ┆            ┆          ┆ ts        ┆   ┆ C.T.b…    ┆         ┆           ┆ 00016…    │\n",
       "│ …          ┆ …          ┆ …        ┆ …         ┆ … ┆ …         ┆ …       ┆ …         ┆ …         │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 567350.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 67350.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11956735 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 0.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 569731.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 69731.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11956973 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 1.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 570829.C ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 70829.C.T │\n",
       "│            ┆            ┆ .T.b38   ┆           ┆   ┆ .11957082 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 9.C.T…    ┆         ┆           ┆ 00000…    │\n",
       "│ ENSG000000 ┆ 4.3164e-24 ┆ chrX.119 ┆ Stomach   ┆ … ┆ GTEXEQTL  ┆ Stomach ┆ 0000945   ┆ chrX.1195 │\n",
       "│ 18610.13   ┆            ┆ 571613.G ┆           ┆   ┆ eQTL.chrX ┆         ┆           ┆ 71613.G.A │\n",
       "│            ┆            ┆ .A.b38   ┆           ┆   ┆ .11957161 ┆         ┆           ┆ .b38.ENSG │\n",
       "│            ┆            ┆          ┆           ┆   ┆ 3.G.A…    ┆         ┆           ┆ 00000…    │\n",
       "└────────────┴────────────┴──────────┴───────────┴───┴───────────┴─────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_all_GTEx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bin P-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create bins w/o having to map bins to the entire df\n",
    "terms_pval = pd.DataFrame()\n",
    "bins = [0,1e-12,1e-11,1e-10,1e-9,1e-8,1e-7,1e-6,1e-5,1e-4,1e-3,.005,.01,.02,.03,.04,.05,.06]\n",
    "\n",
    "b = []\n",
    "for i in range(0,len(bins)-1):\n",
    "    b.append((bins[i],bins[i+1]) )\n",
    "\n",
    "terms_pval['bin'] = b\n",
    "\n",
    "#terms_pval['lowerbound'] = [format(float(i[0]),'.12f') for i in terms_pval['bin']]\n",
    "#terms_pval['upperbound'] = [format(float(i[1]),'.12f') for i in terms_pval['bin']]\n",
    "\n",
    "terms_pval['Term'] = [str(i[0])+','+str(i[1]) for i in terms_pval['bin']]\n",
    "\n",
    "terms_pval['CodeID_PVAL'] = ['PVALUEBINS '+i for i in terms_pval['Term']]\n",
    "terms_pval['rel'] = 'p_value'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create both Pvalue bin nodes (and the relationships to/from GTEX_EQTL nodes) as well as putting actual pvalues in the `value` column. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create p-value nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_namespace</th>\n",
       "      <th>unit</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>value</th>\n",
       "      <th>upperbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PVALUEBINS 0,1e-12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PVALUEBINS 1e-12,1e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PVALUEBINS 1e-11,1e-10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  node_id  node_namespace  unit  node_synonyms  lowerbound   \n",
       "0      PVALUEBINS 0,1e-12             NaN   NaN            NaN         NaN  \\\n",
       "1  PVALUEBINS 1e-12,1e-11             NaN   NaN            NaN         NaN   \n",
       "2  PVALUEBINS 1e-11,1e-10             NaN   NaN            NaN         NaN   \n",
       "\n",
       "   node_definition  node_label  node_dbxrefs  value  upperbound  \n",
       "0              NaN         NaN           NaN    NaN         NaN  \n",
       "1              NaN         NaN           NaN    NaN         NaN  \n",
       "2              NaN         NaN           NaN    NaN         NaN  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_pval = pd.DataFrame(terms_pval['CodeID_PVAL'].drop_duplicates()).rename(columns={'CodeID_PVAL':'node_id'})\n",
    "nodes_pval = fill_missing_cols(nodes_pval)\n",
    "nodes_pval.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_pval['node_id'] = [i.replace(',','.') for i in nodes_pval['node_id']]\n",
    "#nodes_pval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create p-value edges. Use `pl.cut()` to map p_vals to bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "binned_pvals_pl = pl.cut(eqtl_all_GTEx['pval_beta'], bins)\n",
    "\n",
    "eqtl_all_GTEx_saved = eqtl_all_GTEx\n",
    "\n",
    "# get CodeID_gtex and concat with pvalues and the bins we just mapped\n",
    "eqtl_all_GTEx = pl.concat([eqtl_all_GTEx.select(['CodeID_gtex']),binned_pvals_pl],how='horizontal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#eqtl_all_GTEx_df = eqtl_all_GTEx.select(['CodeID_gtex','pval_beta']).to_pandas()\n",
    "#pvals_2_map = list(eqtl_all_GTEx_df['pval_beta'].values)\n",
    "#binned_pvals = pd.cut(pvals_2_map, bins)\n",
    "#eqtl_all_GTEx['binned_pvals'] = binned_pvals.astype(str) # cojvert from categorical to str type\n",
    "#eqtl_all_GTEx.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "SchemaFieldNotFoundError",
     "evalue": "category",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mSchemaFieldNotFoundError\u001b[0m                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# rename and cast to str\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m eqtl_all_GTEx \u001b[38;5;241m=\u001b[39m \u001b[43meqtl_all_GTEx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrename\u001b[49m\u001b[43m(\u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcategory\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbinned_pvals\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\\\n\u001b[1;32m      3\u001b[0m                         \u001b[38;5;241m.\u001b[39mwith_columns(pl\u001b[38;5;241m.\u001b[39mcol(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinned_pvals\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mcast(pl\u001b[38;5;241m.\u001b[39mUtf8, strict\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\u001b[38;5;241m.\u001b[39malias(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbinned_pvals\u001b[39m\u001b[38;5;124m'\u001b[39m))\n",
      "File \u001b[0;32m~/Library/Python/3.10/lib/python/site-packages/polars/dataframe/frame.py:3324\u001b[0m, in \u001b[0;36mDataFrame.rename\u001b[0;34m(self, mapping)\u001b[0m\n\u001b[1;32m   3296\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrename\u001b[39m(\u001b[38;5;28mself\u001b[39m, mapping: \u001b[38;5;28mdict\u001b[39m[\u001b[38;5;28mstr\u001b[39m, \u001b[38;5;28mstr\u001b[39m]) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Self:\n\u001b[1;32m   3297\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   3298\u001b[0m \u001b[38;5;124;03m    Rename column names.\u001b[39;00m\n\u001b[1;32m   3299\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3321\u001b[0m \n\u001b[1;32m   3322\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m   3323\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_from_pydf(\n\u001b[0;32m-> 3324\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlazy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrename\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmapping\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mno_optimization\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39m_df\n\u001b[1;32m   3325\u001b[0m     )\n",
      "File \u001b[0;32m~/Library/Python/3.10/lib/python/site-packages/polars/lazyframe/frame.py:1475\u001b[0m, in \u001b[0;36mLazyFrame.collect\u001b[0;34m(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, no_optimization, slice_pushdown, common_subplan_elimination, streaming)\u001b[0m\n\u001b[1;32m   1464\u001b[0m     common_subplan_elimination \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m   1466\u001b[0m ldf \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_ldf\u001b[38;5;241m.\u001b[39moptimization_toggle(\n\u001b[1;32m   1467\u001b[0m     type_coercion,\n\u001b[1;32m   1468\u001b[0m     predicate_pushdown,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1473\u001b[0m     streaming,\n\u001b[1;32m   1474\u001b[0m )\n\u001b[0;32m-> 1475\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m wrap_df(\u001b[43mldf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n",
      "\u001b[0;31mSchemaFieldNotFoundError\u001b[0m: category"
     ]
    }
   ],
   "source": [
    "# rename and cast to str\n",
    "eqtl_all_GTEx = eqtl_all_GTEx.rename({\"category\": \"binned_pvals\"})\\\n",
    "                        .with_columns(pl.col('binned_pvals').cast(pl.Utf8, strict=False).alias('binned_pvals'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(len(eqtl_all_GTEx))\n",
    "#print(len(eqtl_all_GTEx['CodeID_gtex'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2089174, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>CodeID_gtex</th><th>pval_beta</th><th>break_point</th><th>binned_pvals</th></tr><tr><td>str</td><td>f64</td><td>f64</td><td>str</td></tr></thead><tbody><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>1.1447e-275</td><td>1.0000e-12</td><td>&quot;(0.0, 1.0e-12]…</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0345103</td><td>0.04</td><td>&quot;(0.03, 0.04]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0547082</td><td>0.06</td><td>&quot;(0.05, 0.06]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0547082</td><td>0.06</td><td>&quot;(0.05, 0.06]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0547082</td><td>0.06</td><td>&quot;(0.05, 0.06]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0547082</td><td>0.06</td><td>&quot;(0.05, 0.06]&quot;</td></tr><tr><td>&quot;GTEXEQTL eQTL.…</td><td>0.0547082</td><td>0.06</td><td>&quot;(0.05, 0.06]&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2_089_174, 4)\n",
       "┌───────────────────────────────────┬─────────────┬─────────────┬────────────────┐\n",
       "│ CodeID_gtex                       ┆ pval_beta   ┆ break_point ┆ binned_pvals   │\n",
       "│ ---                               ┆ ---         ┆ ---         ┆ ---            │\n",
       "│ str                               ┆ f64         ┆ f64         ┆ str            │\n",
       "╞═══════════════════════════════════╪═════════════╪═════════════╪════════════════╡\n",
       "│ GTEXEQTL eQTL.chr1.1434243.G.A.b… ┆ 1.1447e-275 ┆ 1.0000e-12  ┆ (0.0, 1.0e-12] │\n",
       "│ GTEXEQTL eQTL.chr1.1497758.C.T.b… ┆ 1.1447e-275 ┆ 1.0000e-12  ┆ (0.0, 1.0e-12] │\n",
       "│ GTEXEQTL eQTL.chr1.1499000.C.A.b… ┆ 1.1447e-275 ┆ 1.0000e-12  ┆ (0.0, 1.0e-12] │\n",
       "│ GTEXEQTL eQTL.chr1.1499128.C.T.b… ┆ 1.1447e-275 ┆ 1.0000e-12  ┆ (0.0, 1.0e-12] │\n",
       "│ …                                 ┆ …           ┆ …           ┆ …              │\n",
       "│ GTEXEQTL eQTL.chrX.119567350.G.A… ┆ 0.0547082   ┆ 0.06        ┆ (0.05, 0.06]   │\n",
       "│ GTEXEQTL eQTL.chrX.119569731.G.A… ┆ 0.0547082   ┆ 0.06        ┆ (0.05, 0.06]   │\n",
       "│ GTEXEQTL eQTL.chrX.119570829.C.T… ┆ 0.0547082   ┆ 0.06        ┆ (0.05, 0.06]   │\n",
       "│ GTEXEQTL eQTL.chrX.119571613.G.A… ┆ 0.0547082   ┆ 0.06        ┆ (0.05, 0.06]   │\n",
       "└───────────────────────────────────┴─────────────┴─────────────┴────────────────┘"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl_all_GTEx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2089174\n",
      "2089169\n"
     ]
    }
   ],
   "source": [
    "print(len(eqtl_all_GTEx))\n",
    "\n",
    "# Filter out pvals >= .05, they are resulting in NaNs\n",
    "eqtl_all_GTEx = eqtl_all_GTEx.filter(pl.col('pval_beta') < 0.05)   \n",
    "\n",
    "print(len(eqtl_all_GTEx))\n",
    "\n",
    "# Filter out pvals >= .05, they are resulting in NaNs\n",
    "#eqtl_all_GTEx_df = eqtl_all_GTEx_df[eqtl_all_GTEx_df['pval_beta'] <.05]\n",
    "#assert not eqtl_all_GTEx_df['binned_pvals'].isna().sum()\n",
    "#assert not 'nan' in eqtl_all_GTEx_df['binned_pvals']\n",
    "\n",
    "assert not eqtl_all_GTEx.filter(pl.col('binned_pvals') == 'nan').shape[0]\n",
    "assert not eqtl_all_GTEx.select(pl.col('binned_pvals').is_null()).sum().to_pandas().values[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTEx_df = eqtl_all_GTEx.to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat and change name so we can merge on this column to get pval code IDs\n",
    "eqtl_all_GTEx_df['binned_pvals'] = eqtl_all_GTEx_df['binned_pvals'].str.replace(']',')')\n",
    "\n",
    "eqtl_all_GTEx_df = eqtl_all_GTEx_df.rename(columns={'binned_pvals':'bin'})\n",
    "\n",
    "eqtl_all_GTEx_df['bin'] =  eqtl_all_GTEx_df['bin'].str.replace('1.0e','1e')\n",
    "eqtl_all_GTEx_df['bin'] =  eqtl_all_GTEx_df['bin'].str.replace('0.0,','0,')\n",
    "\n",
    "terms_pval['bin'] = terms_pval['bin'].astype(str)\n",
    "\n",
    "terms_pval['bin'] = ['(1e-10, 1e-9)' if i == '(1e-10, 1e-09)' else i for i in terms_pval['bin'] ]\n",
    "terms_pval['bin'] = ['(1e-9, 1e-8)' if i == '(1e-09, 1e-08)' else i for i in terms_pval['bin'] ]\n",
    "terms_pval['bin'] =  ['(1e-8, 1e-7)' if i == '(1e-08, 1e-07)' else i for i in terms_pval['bin'] ]\n",
    "terms_pval['bin'] = ['(1e-7, 1e-6)' if i == '(1e-07, 1e-06)' else i for i in terms_pval['bin'] ]\n",
    "terms_pval['bin'] = ['(1e-6, 0.00001)' if i == '(1e-06, 1e-05)' else i for i in terms_pval['bin'] ]\n",
    "terms_pval['bin'] = ['(0.00001, 0.0001)' if i == '(1e-05, 0.0001)' else i for i in terms_pval['bin'] ]\n",
    "# '(1e-10, 1e-09)'  -- '(1e-10, 1e-9)'\n",
    "# '(1e-09, 1e-08)' --  '(1e-9, 1e-8)'\n",
    "# '(1e-08, 1e-07)' -- '(1e-8, 1e-7)'\n",
    "# '(1e-07, 1e-06)' -- '(1e-7, 1e-6)'\n",
    "# '(1e-06, 1e-05)' -- '(1e-6, 0.00001)'\n",
    "# '(1e-05, 0.0001)' -- '(0.00001, 0.0001)'\n",
    "\n",
    "assert not len(set(eqtl_all_GTEx_df['bin']) - set(terms_pval['bin']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge in pval code ids\n",
    "eqtl_all_GTEx_df = pd.merge(eqtl_all_GTEx_df,terms_pval,on='bin',how='left')\n",
    "#eqtl_all_GTEx_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTEx_df['CodeID_PVAL'] = [i.replace(',','.') for i in eqtl_all_GTEx_df['CodeID_PVAL']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTEx_df['predicate'] = 'p_value'\n",
    "\n",
    "pval_edges = eqtl_all_GTEx_df[['CodeID_gtex','predicate','CodeID_PVAL']]\n",
    "pval_edges.columns = ['subject','predicate','object']\n",
    "\n",
    "pval_edges = pval_edges.drop_duplicates()\n",
    "pval_edges = pl.from_pandas(pval_edges)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create eQTL edges file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create HGNC/UBERON code IDs\n",
    "eqtl_all_GTEx = eqtl_all_GTEx_saved\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTExP = eqtl_all_GTEx.to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTExP['uberon_CodeID'] = ['UBERON ' + i if not i.startswith('EFO') else i \n",
    "                                     for i in eqtl_all_GTExP['UBERON_code'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTExP['hgnc_codeID'] = ['HGNC ' + i for i in eqtl_all_GTExP['hgnc_id'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl_all_GTEx = pl.from_pandas(eqtl_all_GTExP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#|eqtl_all_GTEx = eqtl_all_GTEx.with_columns(pl.map([pl.col('hgnc_id')], \n",
    " #                     lambda s: \"HGNC \" +s[0]).alias(\"hgnc_codeID\"))\\\n",
    " #                            .with_columns(pl.map([pl.col('UBERON_code')], \n",
    " #                     lambda s: \"UBERON \" +s[0]).alias(\"uberon_CodeID\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make eqtl-HGNC and eqtl-UBERON edges\n",
    "#### Took out the dbsnp (rsID) edges for now bc the bigger eqtl Data set that we're now using doesn't contain those mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Create predicates (relationship)\n",
    "#eqtl_all_GTEx['gtex_hgnc_rel'] =    'RO_0001025'   #'located in'    old rel:  'eqtl_in_gene'\n",
    "#eqtl_all_GTEx['gtex_uberon_rel'] =  'RO_0001025'   #'located in'    old rel: 'eqtl_in_tissue'\n",
    "LOCATED_IN_RO_PURL = 'http://purl.obolibrary.org/obo/RO_0001025'\n",
    "\n",
    "eqtl_all_GTEx = eqtl_all_GTEx.with_columns(pl.lit('located in').alias('gtex_hgnc_ub_rel'))\n",
    "\n",
    "######## GTEX_EQTL -- HGNC ########\n",
    "edges1 = eqtl_all_GTEx.select(['CodeID_gtex','gtex_hgnc_ub_rel','hgnc_codeID'])\n",
    "\n",
    "# edges1.isna().sum()\n",
    "#edges1 = edges1.dropna().reset_index(drop=True)  # drop rows where HGNC_ID is  NaN\n",
    "\n",
    "######### GTEX_EQTL -- UBERON #######\n",
    "edges2 = eqtl_all_GTEx.select(['CodeID_gtex','gtex_hgnc_ub_rel','uberon_CodeID'])\n",
    "\n",
    "edges1.columns = edges2.columns = ['subject','predicate','object'] \n",
    "     \n",
    "edges_eqtl = pl.concat([edges1,edges2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create GTEX_EQTL nodes\n",
    "### (Dont need to make HGNC or UBERON nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_gtex_eqtl = eqtl_all_GTEx.select(['CodeID_gtex','pval_beta'])   # ,'pval_beta','unit'\n",
    "nodes_gtex_eqtl.columns = ['node_id','value']\n",
    "\n",
    "nodes_eqtl_gtex = fill_missing_cols(nodes_gtex_eqtl)\n",
    "\n",
    "col_order = ['node_id', 'unit', 'node_namespace', 'value','lowerbound', 'node_synonyms',\n",
    " 'node_label', 'node_dbxrefs', 'node_definition', 'upperbound']\n",
    "\n",
    "\n",
    "nodes_pval = nodes_pval[col_order]\n",
    "nodes_eqtl_gtex = nodes_eqtl_gtex.select(col_order)\n",
    "\n",
    "nodes_eqtl = pl.concat([nodes_eqtl_gtex,pl.from_pandas(nodes_pval)]) \n",
    "nodes_eqtl = nodes_eqtl.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create  eQTL to chromosomal ontology edges. \n",
    "### This is not a true 'mapping' beceause we already know where the eQTL is, we just have to assign it to its chlo range which we can do by dividing each location by a certain amount and then assigning the mod to a region\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get just the gtex_eqtl nodes\n",
    "#  name change--confusing, neeed to fix\n",
    "eqtl_nodes = nodes_eqtl.filter(pl.col('node_id').str.starts_with(f'{SAB} eQTL'))\n",
    "eqtl_nodes\n",
    "\n",
    "# get chr and location from node_id field into their own columns\n",
    "# using EQTL_CODE_DELIM here!!!! Defined above. Important to know what char \n",
    "# to split the node_id field by to get the location of the eQTL\n",
    "eqtl_nodes = eqtl_nodes.with_columns(pl.col(\"node_id\").str.split(EQTL_CODE_DELIM).alias('node_id_split'))\\\n",
    "                            .with_columns(pl.col('node_id_split').arr.get(1).alias('chromosome'))\\\n",
    "                            .with_columns(pl.col('node_id_split').arr.get(2).alias('location'))\\\n",
    "                            .with_columns(pl.col('location').cast(pl.Int64, strict=False).alias('location'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "#x = eqtl_nodes.select('node_id').sample(10000).to_pandas()['node_id']\n",
    "\n",
    "#eqtl_nodes = eqtl_nodes.select('node_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "chro_path='/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/HSCLO/OWLNETS_node_metadata.txt'\n",
    "#chro = pl.read_csv(chro_path,separator='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>upperbound</th>\n",
       "      <th>unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHLO chr1.1-1000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHLO chr1.1000001-2000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHLO chr1.2000001-3000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHLO chr1.3000001-4000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHLO chr1.4000001-5000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431148</th>\n",
       "      <td>CHLO MtDNA.5001-6000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431149</th>\n",
       "      <td>CHLO MtDNA.6001-7000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431150</th>\n",
       "      <td>CHLO MtDNA.7001-8000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431151</th>\n",
       "      <td>CHLO MtDNA.8001-9000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431152</th>\n",
       "      <td>CHLO MtDNA.9001-10000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3431153 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           node_id  node_label  node_definition   \n",
       "0              CHLO chr1.1-1000000         NaN              NaN  \\\n",
       "1        CHLO chr1.1000001-2000000         NaN              NaN   \n",
       "2        CHLO chr1.2000001-3000000         NaN              NaN   \n",
       "3        CHLO chr1.3000001-4000000         NaN              NaN   \n",
       "4        CHLO chr1.4000001-5000000         NaN              NaN   \n",
       "...                            ...         ...              ...   \n",
       "3431148       CHLO MtDNA.5001-6000         NaN              NaN   \n",
       "3431149       CHLO MtDNA.6001-7000         NaN              NaN   \n",
       "3431150       CHLO MtDNA.7001-8000         NaN              NaN   \n",
       "3431151       CHLO MtDNA.8001-9000         NaN              NaN   \n",
       "3431152      CHLO MtDNA.9001-10000         NaN              NaN   \n",
       "\n",
       "         node_synonyms node_dbxrefs  value  lowerbound  upperbound  unit  \n",
       "0                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "1                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "2                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "4                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "...                ...          ...    ...         ...         ...   ...  \n",
       "3431148            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431149            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431150            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431151            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431152            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "\n",
       "[3431153 rows x 9 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chro =  pd.read_csv(chro_path,sep='\\t')\n",
    "chro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>upperbound</th>\n",
       "      <th>unit</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HSCLO chr1.1-1000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HSCLO chr1.1000001-2000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HSCLO chr1.2000001-3000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HSCLO chr1.3000001-4000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HSCLO chr1.4000001-5000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431148</th>\n",
       "      <td>HSCLO MtDNA.5001-6000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431149</th>\n",
       "      <td>HSCLO MtDNA.6001-7000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431150</th>\n",
       "      <td>HSCLO MtDNA.7001-8000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431151</th>\n",
       "      <td>HSCLO MtDNA.8001-9000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3431152</th>\n",
       "      <td>HSCLO MtDNA.9001-10000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3431153 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            node_id  node_label  node_definition   \n",
       "0              HSCLO chr1.1-1000000         NaN              NaN  \\\n",
       "1        HSCLO chr1.1000001-2000000         NaN              NaN   \n",
       "2        HSCLO chr1.2000001-3000000         NaN              NaN   \n",
       "3        HSCLO chr1.3000001-4000000         NaN              NaN   \n",
       "4        HSCLO chr1.4000001-5000000         NaN              NaN   \n",
       "...                             ...         ...              ...   \n",
       "3431148       HSCLO MtDNA.5001-6000         NaN              NaN   \n",
       "3431149       HSCLO MtDNA.6001-7000         NaN              NaN   \n",
       "3431150       HSCLO MtDNA.7001-8000         NaN              NaN   \n",
       "3431151       HSCLO MtDNA.8001-9000         NaN              NaN   \n",
       "3431152      HSCLO MtDNA.9001-10000         NaN              NaN   \n",
       "\n",
       "         node_synonyms node_dbxrefs  value  lowerbound  upperbound  unit  \n",
       "0                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "1                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "2                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "4                  NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "...                ...          ...    ...         ...         ...   ...  \n",
       "3431148            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431149            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431150            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431151            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "3431152            NaN          NaN    NaN         NaN         NaN   NaN  \n",
       "\n",
       "[3431153 rows x 9 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chro['node_id'] = [i.replace('CHLO','HSCLO') for i in chro['node_id']]\n",
    "chro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dynamically determine what the `chrom-interval delimiter` (ie. chr1:1-1000, its `:`, but thats been changed to `-`. ) We will count the # of `:`'s in the string. If its 0 then `chrom-interval delimiter` is `-`, it's 1 then `chrom-interval delimiter` is `:`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrom_interval_delimiter is \"-\"\n"
     ]
    }
   ],
   "source": [
    "# Load and format chromosomal ontology dataset\n",
    "chro_path='/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/HSCLO/OWLNETS_node_metadata.txt'\n",
    "chro = pl.read_csv(chro_path,separator='\\t')\n",
    "\n",
    "if chro.select('node_id').head(1000).to_pandas()['node_id'].str.count(':').sum():\n",
    "    print('chrom_interval_delimiter is \":\"')\n",
    "    chrom_interval_delimiter=':'\n",
    "else:\n",
    "    print('chrom_interval_delimiter is \"-\"')\n",
    "    chrom_interval_delimiter='-'\n",
    "    \n",
    "# dont want mtDNA and some rows are chromosome rows that we dont want (no chrom_interval_delimiter in those)\n",
    "chro = chro.filter(~pl.col('node_id').str.starts_with('MtDNA'))\\\n",
    "           .filter(pl.col('node_id').str.contains(chrom_interval_delimiter))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reformat cols so we can isolate chr/start/end into their own cols\n",
    "chro = chro.with_columns(pl.col(\"node_id\").str.split('.')\\\n",
    "                  .arr.to_struct(n_field_strategy=\"max_width\").alias('temp')).unnest('temp')\n",
    "\n",
    "chro.columns = ['node_id','node_label','node_definition','node_synonyms',\n",
    " 'node_dbxrefs','value','lowerbound','upperbound','unit', 'chromosome','intervals']\n",
    "\n",
    "chro = chro.with_columns(pl.col(\"intervals\").str.split(\"-\")\\\n",
    "                  .arr.to_struct(n_field_strategy=\"max_width\").alias('temp')).unnest('temp')      \n",
    "\n",
    "chro.columns = ['node_id','node_label','node_definition','node_synonyms',\n",
    " 'node_dbxrefs','value','lowerbound','upperbound','unit', 'chromosome','intervals','low','high']\n",
    "\n",
    "chro = chro.with_columns(pl.col('chromosome').str.replace('HSCLO ','').alias('chromosome'))\n",
    "\n",
    "chro = chro.with_columns(pl.col('low').cast(pl.Int64, strict=False).alias('low'))\\\n",
    "            .with_columns(pl.col('high').cast(pl.Int64, strict=False).alias('high'))\n",
    "\n",
    "chro = chro.with_columns((pl.col('high') - pl.col('low')).alias('diff'))\n",
    "chro = chro.filter(pl.col('diff') <= 1_000)\n",
    "\n",
    "chro = chro.with_columns(pl.map([pl.col('low'),pl.col('high')],lambda s: s[0]+','+s[1]).alias(\"bin\"))\n",
    "\n",
    "#chro.filter(pl.col('high_mod100')!=0)#.filter(pl.col('chromosome')=='chr1')\n",
    "chro_ends = chro.filter(pl.col('diff')!=999)\n",
    "\n",
    "chro = chro.filter(pl.col('diff')==999) # this gives the rows that contain the ending location of each chrom\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def rounddown(x):\n",
    "#    x100 = x % 1_000 # will round down to nearest nth, whateverr this num is\n",
    "#    #return x if not x100 else x + 10_000 - x100\n",
    "#    return x if not x100 else x - x100\n",
    "\n",
    "# Implement rounddown() in two steps (both column wise so super fast) on the eqtl_nodes dataframe\n",
    "eqtl_nodes = eqtl_nodes.with_columns( (pl.col('location') % 1_000).alias('mod1000'))\n",
    "\n",
    "eqtl_nodes = eqtl_nodes.with_columns(\n",
    "                   pl.when(pl.col(\"mod1000\")==0)\n",
    "                     .then(pl.col('location'))\n",
    "                     .otherwise(pl.col('location') - pl.col(\"mod1000\")).alias('low'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1240810, 3)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "eqtl_nodes = eqtl_nodes.with_columns((pl.col('low') + 1000).alias('high'))\n",
    "\n",
    "eqtl_nodes = eqtl_nodes.with_columns((pl.col('low')+1).alias('low'))\n",
    "\n",
    "eqtl_nodes = eqtl_nodes.with_columns(\n",
    "                        pl.map([pl.col('chromosome'),pl.col('low'),pl.col('high')],\n",
    "                               lambda s: s[0] + '.' + s[1] + '-' + s[2]).alias(\"chro_node_id\"))\n",
    "\n",
    "# ------ quality check ------------------------------------------  # rewrite so it reflects filtering\n",
    "#eqtl_ids = eqtl_nodes.select('chro_node_id').unique()\n",
    "#chro_ids = chro.select('node_id').unique()\n",
    "#eqtl_ids_set = set(list(eqtl_ids.to_pandas()['chro_node_id'].values))\n",
    "#chro_ids_set = set(list(chro_ids.to_pandas()['node_id'].values))\n",
    "# check that all node_ids I made are in the main id list\n",
    "#assert len(eqtl_ids_set & chro_ids_set) == len(eqtl_ids_set)         # uncomment after fixing \n",
    "# ----------------------------------------------------------------\n",
    "\n",
    "# Add the relationship/predicate b/t the eqtl and chromosomal location\n",
    "LOCATED_IN_RO_PURL = 'http://purl.obolibrary.org/obo/RO_0001025'\n",
    "\n",
    "eqtl_nodes = eqtl_nodes.with_columns(pl.lit('located in').alias('predicate'))\n",
    "\n",
    "edges_chrom_ont = eqtl_nodes.select('node_id','predicate','chro_node_id')\n",
    "edges_chrom_ont.columns = ['subject','predicate','object']\n",
    "\n",
    "# Add SAB to chromosome ontology Codes\n",
    "edges_chrom_ont = edges_chrom_ont.with_columns(pl.concat_str('CHLO '+pl.col('object')).alias('object'))\n",
    "\n",
    "# there are many dups here -- bc of multiple eqtls in each tissue -- we need that for the eqtl--tissue edges\n",
    "# but not here, so drop dups.\n",
    "edges_chrom_ont_dd = edges_chrom_ont.unique() # drop duplicates\n",
    "edges_chrom_ont_dd.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concat edges_eqtl (which has eqtl--uberon and eqtl--hgnc edges) with the eqtl--chromosomal_ontology edges and the pval edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5780111, 3)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_order_edges = ['subject','predicate','object']\n",
    "\n",
    "edges_eqtl = edges_eqtl.select(col_order_edges)\n",
    "edges_chrom_ont_dd = edges_chrom_ont_dd.select(col_order_edges)\n",
    "pval_edges = pval_edges.select(col_order_edges)\n",
    "\n",
    "# character to replace the colon with in the code, ie chr:1-1000\n",
    "\n",
    "# make sure these chlo nodes match the ones below (check the format/delim is the same)\n",
    "edges_chrom_ont_dd = edges_chrom_ont_dd.with_columns(pl.col('object').str.replace(':','.').alias('object'))\n",
    "edges_eqtl = pl.concat([edges_eqtl,edges_chrom_ont_dd,pval_edges])\n",
    "\n",
    "edges_eqtl = edges_eqtl.unique()\n",
    "\n",
    "edges_eqtl.unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add CHLO nodeds to the eqtl edges so we can ingest GTEX_EQTL data w/o needing CHLO already in the graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "chro_path='/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/HSCLO/OWLNETS_node_metadata.txt'\n",
    "chro = pl.read_csv(chro_path,separator='\\t')\n",
    "chro = fill_missing_cols(chro).select(col_order)\n",
    "\n",
    "# make sure these chlo nodes match the ones above (check the format/delim is the same)\n",
    "chro = chro.with_columns(pl.col('node_id').str.replace(':','.').alias('node_id'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### check that CHLO codes in `edges_chrom_ont_dd.select('object')` match those in `chro.select('node_id')`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert not len(set([i[0] for i in edges_chrom_ont_dd.select('object').to_pandas().values]) -\\\n",
    "set([i[0] for i in chro.select('node_id').to_pandas().values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_all = pd.concat([nodes_eqtl.to_pandas(),chro.to_pandas()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edges_eqtl.filter(pl.col('object').str.contains('PVALUE')).select('object').head().to_pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_all = nodes_all.drop_duplicates()\n",
    "edges_eqtl = edges_eqtl.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#edges_eqtl.select(pl.col('subject').str.startswith('CHLO'))\n",
    "#eedf = edges_eqtl.to_pandas()\n",
    "\n",
    "#eedf[eedf['object'].str.startswith('CHLO')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#nodes_all[nodes_all['node_id'].str.startswith('CHLO')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving nodes and edges...\n"
     ]
    }
   ],
   "source": [
    "print('saving nodes and edges...')\n",
    "\n",
    "nodes.to_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_eqtl/OWLNETS_node_metadata.txt',\n",
    "             sep='\\t',index=False)\n",
    "\n",
    "#edges_eqtl.to_pandas().to_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_eqtl/OWLNETS_edgelist.txt',\n",
    "#             sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_eqtl = pd.read_csv(\n",
    "    '/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_eqtl/OWLNETS_edgelist.txt',sep='\\t')\n",
    "#edges_eqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [subject, predicate, object]\n",
       "Index: []"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_eqtl[edges_eqtl['object'].str.startswith('CHLO')]\n",
    "\n",
    "#edges_eqtl['object'] = [i.replace('CHLO','HSCLO') for i in edges_eqtl['object']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_eqtl.to_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_eqtl/OWLNETS_edgelist.txt',\n",
    "             sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>unit</th>\n",
       "      <th>node_namespace</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>upperbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478253</th>\n",
       "      <td>CHLO MtDNA.5001-6000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478254</th>\n",
       "      <td>CHLO MtDNA.6001-7000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478255</th>\n",
       "      <td>CHLO MtDNA.7001-8000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478256</th>\n",
       "      <td>CHLO MtDNA.8001-9000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478257</th>\n",
       "      <td>CHLO MtDNA.9001-10000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5478258 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               node_id  unit   \n",
       "0        GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts   NaN  \\\n",
       "1        GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "2        GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "3        GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "4        GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "...                                                                ...   ...   \n",
       "5478253                                           CHLO MtDNA.5001-6000   NaN   \n",
       "5478254                                           CHLO MtDNA.6001-7000   NaN   \n",
       "5478255                                           CHLO MtDNA.7001-8000   NaN   \n",
       "5478256                                           CHLO MtDNA.8001-9000   NaN   \n",
       "5478257                                          CHLO MtDNA.9001-10000   NaN   \n",
       "\n",
       "         node_namespace         value  lowerbound  node_synonyms  node_label   \n",
       "0                   NaN  5.686160e-41         NaN            NaN         NaN  \\\n",
       "1                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "2                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "3                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "4                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "...                 ...           ...         ...            ...         ...   \n",
       "5478253             NaN           NaN         NaN            NaN         NaN   \n",
       "5478254             NaN           NaN         NaN            NaN         NaN   \n",
       "5478255             NaN           NaN         NaN            NaN         NaN   \n",
       "5478256             NaN           NaN         NaN            NaN         NaN   \n",
       "5478257             NaN           NaN         NaN            NaN         NaN   \n",
       "\n",
       "        node_dbxrefs  node_definition  upperbound  \n",
       "0                NaN              NaN         NaN  \n",
       "1                NaN              NaN         NaN  \n",
       "2                NaN              NaN         NaN  \n",
       "3                NaN              NaN         NaN  \n",
       "4                NaN              NaN         NaN  \n",
       "...              ...              ...         ...  \n",
       "5478253          NaN              NaN         NaN  \n",
       "5478254          NaN              NaN         NaN  \n",
       "5478255          NaN              NaN         NaN  \n",
       "5478256          NaN              NaN         NaN  \n",
       "5478257          NaN              NaN         NaN  \n",
       "\n",
       "[5478258 rows x 10 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes = pd.read_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_eqtl/OWLNETS_node_metadata.txt',\n",
    "             sep='\\t')\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>unit</th>\n",
       "      <th>node_namespace</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>upperbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478253</th>\n",
       "      <td>CHLO MtDNA.5001-6000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478254</th>\n",
       "      <td>CHLO MtDNA.6001-7000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478255</th>\n",
       "      <td>CHLO MtDNA.7001-8000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478256</th>\n",
       "      <td>CHLO MtDNA.8001-9000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478257</th>\n",
       "      <td>CHLO MtDNA.9001-10000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5478258 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               node_id  unit   \n",
       "0        GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts   NaN  \\\n",
       "1        GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "2        GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "3        GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "4        GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "...                                                                ...   ...   \n",
       "5478253                                           CHLO MtDNA.5001-6000   NaN   \n",
       "5478254                                           CHLO MtDNA.6001-7000   NaN   \n",
       "5478255                                           CHLO MtDNA.7001-8000   NaN   \n",
       "5478256                                           CHLO MtDNA.8001-9000   NaN   \n",
       "5478257                                          CHLO MtDNA.9001-10000   NaN   \n",
       "\n",
       "         node_namespace         value  lowerbound  node_synonyms  node_label   \n",
       "0                   NaN  5.686160e-41         NaN            NaN         NaN  \\\n",
       "1                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "2                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "3                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "4                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "...                 ...           ...         ...            ...         ...   \n",
       "5478253             NaN           NaN         NaN            NaN         NaN   \n",
       "5478254             NaN           NaN         NaN            NaN         NaN   \n",
       "5478255             NaN           NaN         NaN            NaN         NaN   \n",
       "5478256             NaN           NaN         NaN            NaN         NaN   \n",
       "5478257             NaN           NaN         NaN            NaN         NaN   \n",
       "\n",
       "        node_dbxrefs  node_definition  upperbound  \n",
       "0                NaN              NaN         NaN  \n",
       "1                NaN              NaN         NaN  \n",
       "2                NaN              NaN         NaN  \n",
       "3                NaN              NaN         NaN  \n",
       "4                NaN              NaN         NaN  \n",
       "...              ...              ...         ...  \n",
       "5478253          NaN              NaN         NaN  \n",
       "5478254          NaN              NaN         NaN  \n",
       "5478255          NaN              NaN         NaN  \n",
       "5478256          NaN              NaN         NaN  \n",
       "5478257          NaN              NaN         NaN  \n",
       "\n",
       "[5478258 rows x 10 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>unit</th>\n",
       "      <th>node_namespace</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>node_synonyms</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>upperbound</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.686160e-41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478253</th>\n",
       "      <td>HSCLO MtDNA.5001-6000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478254</th>\n",
       "      <td>HSCLO MtDNA.6001-7000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478255</th>\n",
       "      <td>HSCLO MtDNA.7001-8000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478256</th>\n",
       "      <td>HSCLO MtDNA.8001-9000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5478257</th>\n",
       "      <td>HSCLO MtDNA.9001-10000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5478258 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               node_id  unit   \n",
       "0        GTEXEQTL eQTL.chr1.1434243.G.A.b38.Cells.Cultured.fibroblasts   NaN  \\\n",
       "1        GTEXEQTL eQTL.chr1.1497758.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "2        GTEXEQTL eQTL.chr1.1499000.C.A.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "3        GTEXEQTL eQTL.chr1.1499128.C.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "4        GTEXEQTL eQTL.chr1.1499639.G.T.b38.Cells.Cultured.fibroblasts   NaN   \n",
       "...                                                                ...   ...   \n",
       "5478253                                          HSCLO MtDNA.5001-6000   NaN   \n",
       "5478254                                          HSCLO MtDNA.6001-7000   NaN   \n",
       "5478255                                          HSCLO MtDNA.7001-8000   NaN   \n",
       "5478256                                          HSCLO MtDNA.8001-9000   NaN   \n",
       "5478257                                         HSCLO MtDNA.9001-10000   NaN   \n",
       "\n",
       "         node_namespace         value  lowerbound  node_synonyms  node_label   \n",
       "0                   NaN  5.686160e-41         NaN            NaN         NaN  \\\n",
       "1                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "2                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "3                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "4                   NaN  5.686160e-41         NaN            NaN         NaN   \n",
       "...                 ...           ...         ...            ...         ...   \n",
       "5478253             NaN           NaN         NaN            NaN         NaN   \n",
       "5478254             NaN           NaN         NaN            NaN         NaN   \n",
       "5478255             NaN           NaN         NaN            NaN         NaN   \n",
       "5478256             NaN           NaN         NaN            NaN         NaN   \n",
       "5478257             NaN           NaN         NaN            NaN         NaN   \n",
       "\n",
       "        node_dbxrefs  node_definition  upperbound  \n",
       "0                NaN              NaN         NaN  \n",
       "1                NaN              NaN         NaN  \n",
       "2                NaN              NaN         NaN  \n",
       "3                NaN              NaN         NaN  \n",
       "4                NaN              NaN         NaN  \n",
       "...              ...              ...         ...  \n",
       "5478253          NaN              NaN         NaN  \n",
       "5478254          NaN              NaN         NaN  \n",
       "5478255          NaN              NaN         NaN  \n",
       "5478256          NaN              NaN         NaN  \n",
       "5478257          NaN              NaN         NaN  \n",
       "\n",
       "[5478258 rows x 10 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes['node_id'] = [i.replace('CHLO','HSCLO') for i in nodes['node_id']]\n",
    "nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now let's look at the Gene Median TPM dataset\n",
    "File: GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct  \n",
    "Description: This file contains genes as rows And tissues as columns. Unlike the other gene expression GTEx dataset we looked at where columns were individual samples (of which we had 17,382 samples), the gene median dataset has only 54 columns corresponding to each tissue type. This gene median data set shows just the median level expression for each tissue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(56200, 54)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>cid</th>\n",
       "      <th>Adipose - Subcutaneous</th>\n",
       "      <th>Adipose - Visceral (Omentum)</th>\n",
       "      <th>Adrenal Gland</th>\n",
       "      <th>Artery - Aorta</th>\n",
       "      <th>Artery - Coronary</th>\n",
       "      <th>Artery - Tibial</th>\n",
       "      <th>Bladder</th>\n",
       "      <th>Brain - Amygdala</th>\n",
       "      <th>Brain - Anterior cingulate cortex (BA24)</th>\n",
       "      <th>Brain - Caudate (basal ganglia)</th>\n",
       "      <th>Brain - Cerebellar Hemisphere</th>\n",
       "      <th>Brain - Cerebellum</th>\n",
       "      <th>Brain - Cortex</th>\n",
       "      <th>Brain - Frontal Cortex (BA9)</th>\n",
       "      <th>Brain - Hippocampus</th>\n",
       "      <th>Brain - Hypothalamus</th>\n",
       "      <th>Brain - Nucleus accumbens (basal ganglia)</th>\n",
       "      <th>Brain - Putamen (basal ganglia)</th>\n",
       "      <th>Brain - Spinal cord (cervical c-1)</th>\n",
       "      <th>Brain - Substantia nigra</th>\n",
       "      <th>Breast - Mammary Tissue</th>\n",
       "      <th>Cells - Cultured fibroblasts</th>\n",
       "      <th>Cells - EBV-transformed lymphocytes</th>\n",
       "      <th>Cervix - Ectocervix</th>\n",
       "      <th>Cervix - Endocervix</th>\n",
       "      <th>Colon - Sigmoid</th>\n",
       "      <th>Colon - Transverse</th>\n",
       "      <th>Esophagus - Gastroesophageal Junction</th>\n",
       "      <th>Esophagus - Mucosa</th>\n",
       "      <th>Esophagus - Muscularis</th>\n",
       "      <th>Fallopian Tube</th>\n",
       "      <th>Heart - Atrial Appendage</th>\n",
       "      <th>Heart - Left Ventricle</th>\n",
       "      <th>Kidney - Cortex</th>\n",
       "      <th>Kidney - Medulla</th>\n",
       "      <th>Liver</th>\n",
       "      <th>Lung</th>\n",
       "      <th>Minor Salivary Gland</th>\n",
       "      <th>Muscle - Skeletal</th>\n",
       "      <th>Nerve - Tibial</th>\n",
       "      <th>Ovary</th>\n",
       "      <th>Pancreas</th>\n",
       "      <th>Pituitary</th>\n",
       "      <th>Prostate</th>\n",
       "      <th>Skin - Not Sun Exposed (Suprapubic)</th>\n",
       "      <th>Skin - Sun Exposed (Lower leg)</th>\n",
       "      <th>Small Intestine - Terminal Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole Blood</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.166403</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>4.06403</td>\n",
       "      <td>3.37111</td>\n",
       "      <td>2.68549</td>\n",
       "      <td>4.04762</td>\n",
       "      <td>3.90076</td>\n",
       "      <td>3.63963</td>\n",
       "      <td>5.16375</td>\n",
       "      <td>1.43859</td>\n",
       "      <td>1.69285</td>\n",
       "      <td>1.56605</td>\n",
       "      <td>4.99231</td>\n",
       "      <td>5.72099</td>\n",
       "      <td>2.48317</td>\n",
       "      <td>2.14667</td>\n",
       "      <td>1.68599</td>\n",
       "      <td>1.74811</td>\n",
       "      <td>1.53899</td>\n",
       "      <td>1.44167</td>\n",
       "      <td>2.73049</td>\n",
       "      <td>1.74194</td>\n",
       "      <td>4.43876</td>\n",
       "      <td>1.6786</td>\n",
       "      <td>2.49477</td>\n",
       "      <td>5.62935</td>\n",
       "      <td>7.09749</td>\n",
       "      <td>4.64777</td>\n",
       "      <td>3.59509</td>\n",
       "      <td>4.32641</td>\n",
       "      <td>3.11749</td>\n",
       "      <td>4.10335</td>\n",
       "      <td>6.13409</td>\n",
       "      <td>1.52031</td>\n",
       "      <td>0.924962</td>\n",
       "      <td>2.77081</td>\n",
       "      <td>2.21451</td>\n",
       "      <td>1.76541</td>\n",
       "      <td>4.50841</td>\n",
       "      <td>3.52767</td>\n",
       "      <td>1.41667</td>\n",
       "      <td>6.68531</td>\n",
       "      <td>6.6341</td>\n",
       "      <td>1.80871</td>\n",
       "      <td>5.42546</td>\n",
       "      <td>7.08318</td>\n",
       "      <td>5.93298</td>\n",
       "      <td>6.13265</td>\n",
       "      <td>4.19378</td>\n",
       "      <td>5.92631</td>\n",
       "      <td>3.06248</td>\n",
       "      <td>4.702530</td>\n",
       "      <td>6.27255</td>\n",
       "      <td>7.19001</td>\n",
       "      <td>5.74554</td>\n",
       "      <td>2.64743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "cid                Adipose - Subcutaneous  Adipose - Visceral (Omentum)   \n",
       "rid                                                                       \n",
       "ENSG00000223972.5                 0.00000                       0.00000  \\\n",
       "ENSG00000227232.5                 4.06403                       3.37111   \n",
       "ENSG00000278267.1                 0.00000                       0.00000   \n",
       "\n",
       "cid                Adrenal Gland  Artery - Aorta  Artery - Coronary   \n",
       "rid                                                                   \n",
       "ENSG00000223972.5        0.00000         0.00000            0.00000  \\\n",
       "ENSG00000227232.5        2.68549         4.04762            3.90076   \n",
       "ENSG00000278267.1        0.00000         0.00000            0.00000   \n",
       "\n",
       "cid                Artery - Tibial  Bladder  Brain - Amygdala   \n",
       "rid                                                             \n",
       "ENSG00000223972.5          0.00000  0.00000           0.00000  \\\n",
       "ENSG00000227232.5          3.63963  5.16375           1.43859   \n",
       "ENSG00000278267.1          0.00000  0.00000           0.00000   \n",
       "\n",
       "cid                Brain - Anterior cingulate cortex (BA24)   \n",
       "rid                                                           \n",
       "ENSG00000223972.5                                   0.00000  \\\n",
       "ENSG00000227232.5                                   1.69285   \n",
       "ENSG00000278267.1                                   0.00000   \n",
       "\n",
       "cid                Brain - Caudate (basal ganglia)   \n",
       "rid                                                  \n",
       "ENSG00000223972.5                          0.00000  \\\n",
       "ENSG00000227232.5                          1.56605   \n",
       "ENSG00000278267.1                          0.00000   \n",
       "\n",
       "cid                Brain - Cerebellar Hemisphere  Brain - Cerebellum   \n",
       "rid                                                                    \n",
       "ENSG00000223972.5                        0.00000             0.00000  \\\n",
       "ENSG00000227232.5                        4.99231             5.72099   \n",
       "ENSG00000278267.1                        0.00000             0.00000   \n",
       "\n",
       "cid                Brain - Cortex  Brain - Frontal Cortex (BA9)   \n",
       "rid                                                               \n",
       "ENSG00000223972.5         0.00000                       0.00000  \\\n",
       "ENSG00000227232.5         2.48317                       2.14667   \n",
       "ENSG00000278267.1         0.00000                       0.00000   \n",
       "\n",
       "cid                Brain - Hippocampus  Brain - Hypothalamus   \n",
       "rid                                                            \n",
       "ENSG00000223972.5              0.00000               0.00000  \\\n",
       "ENSG00000227232.5              1.68599               1.74811   \n",
       "ENSG00000278267.1              0.00000               0.00000   \n",
       "\n",
       "cid                Brain - Nucleus accumbens (basal ganglia)   \n",
       "rid                                                            \n",
       "ENSG00000223972.5                                    0.00000  \\\n",
       "ENSG00000227232.5                                    1.53899   \n",
       "ENSG00000278267.1                                    0.00000   \n",
       "\n",
       "cid                Brain - Putamen (basal ganglia)   \n",
       "rid                                                  \n",
       "ENSG00000223972.5                          0.00000  \\\n",
       "ENSG00000227232.5                          1.44167   \n",
       "ENSG00000278267.1                          0.00000   \n",
       "\n",
       "cid                Brain - Spinal cord (cervical c-1)   \n",
       "rid                                                     \n",
       "ENSG00000223972.5                             0.00000  \\\n",
       "ENSG00000227232.5                             2.73049   \n",
       "ENSG00000278267.1                             0.00000   \n",
       "\n",
       "cid                Brain - Substantia nigra  Breast - Mammary Tissue   \n",
       "rid                                                                    \n",
       "ENSG00000223972.5                   0.00000                  0.00000  \\\n",
       "ENSG00000227232.5                   1.74194                  4.43876   \n",
       "ENSG00000278267.1                   0.00000                  0.00000   \n",
       "\n",
       "cid                Cells - Cultured fibroblasts   \n",
       "rid                                               \n",
       "ENSG00000223972.5                        0.0000  \\\n",
       "ENSG00000227232.5                        1.6786   \n",
       "ENSG00000278267.1                        0.0000   \n",
       "\n",
       "cid                Cells - EBV-transformed lymphocytes  Cervix - Ectocervix   \n",
       "rid                                                                           \n",
       "ENSG00000223972.5                              0.00000              0.00000  \\\n",
       "ENSG00000227232.5                              2.49477              5.62935   \n",
       "ENSG00000278267.1                              0.00000              0.00000   \n",
       "\n",
       "cid                Cervix - Endocervix  Colon - Sigmoid  Colon - Transverse   \n",
       "rid                                                                           \n",
       "ENSG00000223972.5              0.00000          0.00000             0.00000  \\\n",
       "ENSG00000227232.5              7.09749          4.64777             3.59509   \n",
       "ENSG00000278267.1              0.00000          0.00000             0.00000   \n",
       "\n",
       "cid                Esophagus - Gastroesophageal Junction  Esophagus - Mucosa   \n",
       "rid                                                                            \n",
       "ENSG00000223972.5                                0.00000             0.00000  \\\n",
       "ENSG00000227232.5                                4.32641             3.11749   \n",
       "ENSG00000278267.1                                0.00000             0.00000   \n",
       "\n",
       "cid                Esophagus - Muscularis  Fallopian Tube   \n",
       "rid                                                         \n",
       "ENSG00000223972.5                 0.00000         0.00000  \\\n",
       "ENSG00000227232.5                 4.10335         6.13409   \n",
       "ENSG00000278267.1                 0.00000         0.00000   \n",
       "\n",
       "cid                Heart - Atrial Appendage  Heart - Left Ventricle   \n",
       "rid                                                                   \n",
       "ENSG00000223972.5                   0.00000                0.000000  \\\n",
       "ENSG00000227232.5                   1.52031                0.924962   \n",
       "ENSG00000278267.1                   0.00000                0.000000   \n",
       "\n",
       "cid                Kidney - Cortex  Kidney - Medulla    Liver     Lung   \n",
       "rid                                                                      \n",
       "ENSG00000223972.5          0.00000           0.00000  0.00000  0.00000  \\\n",
       "ENSG00000227232.5          2.77081           2.21451  1.76541  4.50841   \n",
       "ENSG00000278267.1          0.00000           0.00000  0.00000  0.00000   \n",
       "\n",
       "cid                Minor Salivary Gland  Muscle - Skeletal  Nerve - Tibial   \n",
       "rid                                                                          \n",
       "ENSG00000223972.5               0.00000            0.00000         0.00000  \\\n",
       "ENSG00000227232.5               3.52767            1.41667         6.68531   \n",
       "ENSG00000278267.1               0.00000            0.00000         0.00000   \n",
       "\n",
       "cid                 Ovary  Pancreas  Pituitary  Prostate   \n",
       "rid                                                        \n",
       "ENSG00000223972.5  0.0000   0.00000    0.00000   0.00000  \\\n",
       "ENSG00000227232.5  6.6341   1.80871    5.42546   7.08318   \n",
       "ENSG00000278267.1  0.0000   0.00000    0.00000   0.00000   \n",
       "\n",
       "cid                Skin - Not Sun Exposed (Suprapubic)   \n",
       "rid                                                      \n",
       "ENSG00000223972.5                              0.00000  \\\n",
       "ENSG00000227232.5                              5.93298   \n",
       "ENSG00000278267.1                              0.00000   \n",
       "\n",
       "cid                Skin - Sun Exposed (Lower leg)   \n",
       "rid                                                 \n",
       "ENSG00000223972.5                         0.00000  \\\n",
       "ENSG00000227232.5                         6.13265   \n",
       "ENSG00000278267.1                         0.00000   \n",
       "\n",
       "cid                Small Intestine - Terminal Ileum   Spleen  Stomach   \n",
       "rid                                                                     \n",
       "ENSG00000223972.5                           0.00000  0.00000  0.00000  \\\n",
       "ENSG00000227232.5                           4.19378  5.92631  3.06248   \n",
       "ENSG00000278267.1                           0.00000  0.00000  0.00000   \n",
       "\n",
       "cid                  Testis  Thyroid   Uterus   Vagina  Whole Blood  \n",
       "rid                                                                  \n",
       "ENSG00000223972.5  0.166403  0.00000  0.00000  0.00000      0.00000  \n",
       "ENSG00000227232.5  4.702530  6.27255  7.19001  5.74554      2.64743  \n",
       "ENSG00000278267.1  0.000000  0.00000  0.00000  0.00000      0.00000  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_median_tpm = '/Users/stearb/desktop/DESKTOP_TRANSFER/R03_local/data/gtex/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct'\n",
    "\n",
    "gct_obj=parse(gene_median_tpm)\n",
    "\n",
    "df = gct_obj.data_df\n",
    "print(df.shape)\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show names of the fields for the object\n",
    "#gct_obj.__dict__.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show row (gene) metadata (Column (tissue) metadata is just the column names)\n",
    "#gct_obj.row_metadata_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3034800, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>Median_TPM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>Adipose Subcutaneous</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>Adipose Visceral (Omentum)</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>Adrenal Gland</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Transcript ID                      tissue  Median_TPM\n",
       "0  ENSG00000223972.5        Adipose Subcutaneous         0.0\n",
       "1  ENSG00000223972.5  Adipose Visceral (Omentum)         0.0\n",
       "2  ENSG00000223972.5               Adrenal Gland         0.0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reformat and rename both axes, also use stack to flatten the dataframe (make tissues into a column)\n",
    "medgene_flat = df.stack().reset_index().rename(columns={'rid':'Transcript ID',\n",
    "                                                        'cid':'tissue',0:'Median_TPM'})\n",
    "\n",
    "# Reformat tissue strings so they match the UBERON mappings data\n",
    "medgene_flat['tissue'] = medgene_flat['tissue'].str.replace(' - ',' ').str.replace('\\(','').str.replace('\\)','')\n",
    "print(medgene_flat.shape)\n",
    "medgene_flat.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Look at row (gene) metadata and map them to HGNC IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of unique genes: 54592 (some overlap)\n",
      "# of unique transcript IDs: 56200 (no overlap)\n"
     ]
    }
   ],
   "source": [
    "gtx_genes = gct_obj.row_metadata_df['Description'].to_frame()\n",
    "gtx_genes['Transcript ID'] = gtx_genes.index\n",
    "gtx_genes.rename(columns={'Description':'symbol'},inplace=True)\n",
    "gtx_genes.reset_index(drop=True, inplace=True)\n",
    "\n",
    "print(f'# of unique genes: {gct_obj.row_metadata_df[\"Description\"].nunique()} (some overlap)')\n",
    "print(f'# of unique transcript IDs: {gtx_genes.index.nunique()} (no overlap)')\n",
    "#gtx_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>Transcript ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56195</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56196</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56197</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56198</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56199</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56200 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            symbol      Transcript ID\n",
       "0          DDX11L1  ENSG00000223972.5\n",
       "1           WASH7P  ENSG00000227232.5\n",
       "2        MIR6859-1  ENSG00000278267.1\n",
       "3      MIR1302-2HG  ENSG00000243485.5\n",
       "4          FAM138A  ENSG00000237613.2\n",
       "...            ...                ...\n",
       "56195       MT-ND6  ENSG00000198695.2\n",
       "56196        MT-TE  ENSG00000210194.1\n",
       "56197       MT-CYB  ENSG00000198727.2\n",
       "56198        MT-TT  ENSG00000210195.2\n",
       "56199        MT-TP  ENSG00000210196.2\n",
       "\n",
       "[56200 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtx_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how much overlap is there with the HGNC master genes list and the gtx_genes?\n",
    "#venn2([set(gtx_genes['symbol']),\n",
    "#       set(hgnc_master['symbol'])],\n",
    "#       set_labels = ('GTEx genes', 'HGNC master')); plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use the master hgnc list from the website and filter by gtx_genes so we only get the overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>status</th>\n",
       "      <th>location</th>\n",
       "      <th>location_sortable</th>\n",
       "      <th>alias_symbol</th>\n",
       "      <th>alias_name</th>\n",
       "      <th>prev_symbol</th>\n",
       "      <th>prev_name</th>\n",
       "      <th>gene_group</th>\n",
       "      <th>gene_group_id</th>\n",
       "      <th>date_approved_reserved</th>\n",
       "      <th>date_symbol_changed</th>\n",
       "      <th>date_name_changed</th>\n",
       "      <th>date_modified</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>vega_id</th>\n",
       "      <th>ucsc_id</th>\n",
       "      <th>ena</th>\n",
       "      <th>refseq_accession</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>mgd_id</th>\n",
       "      <th>rgd_id</th>\n",
       "      <th>lsdb</th>\n",
       "      <th>cosmic</th>\n",
       "      <th>omim_id</th>\n",
       "      <th>mirbase</th>\n",
       "      <th>homeodb</th>\n",
       "      <th>snornabase</th>\n",
       "      <th>bioparadigms_slc</th>\n",
       "      <th>orphanet</th>\n",
       "      <th>pseudogene.org</th>\n",
       "      <th>horde_id</th>\n",
       "      <th>merops</th>\n",
       "      <th>imgt</th>\n",
       "      <th>iuphar</th>\n",
       "      <th>kznf_gene_catalog</th>\n",
       "      <th>mamit-trnadb</th>\n",
       "      <th>cd</th>\n",
       "      <th>lncrnadb</th>\n",
       "      <th>enzyme_id</th>\n",
       "      <th>intermediate_filament_db</th>\n",
       "      <th>rna_central_ids</th>\n",
       "      <th>lncipedia</th>\n",
       "      <th>gtrnadb</th>\n",
       "      <th>agr</th>\n",
       "      <th>mane_select</th>\n",
       "      <th>gencc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Immunoglobulin like domain containing</td>\n",
       "      <td>594</td>\n",
       "      <td>1989-06-30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>OTTHUMG00000183507</td>\n",
       "      <td>uc002qsd.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>CCDS12976</td>\n",
       "      <td>P04217</td>\n",
       "      <td>2591067</td>\n",
       "      <td>MGI:2152878</td>\n",
       "      <td>RGD:69417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I43.950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:5</td>\n",
       "      <td>ENST00000263100.8|NM_130786.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCRNA00181|A1BGAS|A1BG-AS</td>\n",
       "      <td>non-protein coding RNA 181|A1BG antisense RNA (non-protein coding)|A1BG antisense RNA 1 (non-protein coding)</td>\n",
       "      <td>Antisense RNAs</td>\n",
       "      <td>1987</td>\n",
       "      <td>2009-07-20</td>\n",
       "      <td>2010-11-25</td>\n",
       "      <td>2012-08-15</td>\n",
       "      <td>2013-06-27</td>\n",
       "      <td>503538.0</td>\n",
       "      <td>ENSG00000268895</td>\n",
       "      <td>OTTHUMG00000183508</td>\n",
       "      <td>uc002qse.3</td>\n",
       "      <td>BC040926</td>\n",
       "      <td>NR_015380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:37133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>ACF|ASP|ACF64|ACF65|APOBEC1CF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA binding motif containing</td>\n",
       "      <td>725</td>\n",
       "      <td>2007-11-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>29974.0</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>OTTHUMG00000018240</td>\n",
       "      <td>uc057tgv.1</td>\n",
       "      <td>AF271790</td>\n",
       "      <td>NM_014576</td>\n",
       "      <td>CCDS7242|CCDS7241|CCDS73133|CCDS7243</td>\n",
       "      <td>Q9NQ94</td>\n",
       "      <td>11815617|11072063</td>\n",
       "      <td>MGI:1917115</td>\n",
       "      <td>RGD:619834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>618199</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24086</td>\n",
       "      <td>ENST00000373997.8|NM_014576.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>FWP007|S863-7|CPAMD5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alpha-2-macroglobulin family</td>\n",
       "      <td>2148</td>\n",
       "      <td>1986-01-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>OTTHUMG00000150267</td>\n",
       "      <td>uc001qvk.2</td>\n",
       "      <td>BX647329|X68728|M11313</td>\n",
       "      <td>NM_000014</td>\n",
       "      <td>CCDS44827</td>\n",
       "      <td>P01023</td>\n",
       "      <td>2408344|9697696</td>\n",
       "      <td>MGI:2449119</td>\n",
       "      <td>RGD:2004</td>\n",
       "      <td>LRG_591|http://ftp.ebi.ac.uk/pub/databases/lrgex/LRG_591.xml</td>\n",
       "      <td>NaN</td>\n",
       "      <td>103950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I39.001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:7</td>\n",
       "      <td>ENST00000318602.12|NM_000014.6</td>\n",
       "      <td>HGNC:7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>A2M antisense RNA 1</td>\n",
       "      <td>non-coding RNA</td>\n",
       "      <td>RNA, long non-coding</td>\n",
       "      <td>Approved</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2M antisense RNA 1 (non-protein coding)|A2M antisense RNA 1</td>\n",
       "      <td>Antisense RNAs</td>\n",
       "      <td>1987</td>\n",
       "      <td>2012-06-23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2018-03-21</td>\n",
       "      <td>2018-03-21</td>\n",
       "      <td>144571.0</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>OTTHUMG00000168289</td>\n",
       "      <td>uc009zgj.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NR_026971</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A2M-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:27057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43648</th>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>zyg-11 family member B, cell cycle regulator</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p32.3</td>\n",
       "      <td>01p32.3</td>\n",
       "      <td>FLJ13456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZYG11</td>\n",
       "      <td>zyg-11 homolog (C. elegans)|zyg-11 homolog B (C. elegans)</td>\n",
       "      <td>ZYG11 cell cycle regulator family|Armadillo like helical domain containing</td>\n",
       "      <td>6|1492</td>\n",
       "      <td>2005-06-06</td>\n",
       "      <td>2005-07-11</td>\n",
       "      <td>2012-12-10</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>79699.0</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>OTTHUMG00000008938</td>\n",
       "      <td>uc001cuj.4</td>\n",
       "      <td>AB051517</td>\n",
       "      <td>NM_024646</td>\n",
       "      <td>CCDS30717</td>\n",
       "      <td>Q9C0D3</td>\n",
       "      <td>11214970</td>\n",
       "      <td>MGI:2685277</td>\n",
       "      <td>RGD:1307814</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>618673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:25820</td>\n",
       "      <td>ENST00000294353.7|NM_024646.3</td>\n",
       "      <td>HGNC:25820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43649</th>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ZYX</td>\n",
       "      <td>zyxin</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>7q34</td>\n",
       "      <td>07q34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zyxin family|MicroRNA protein coding host genes</td>\n",
       "      <td>1402|1691</td>\n",
       "      <td>1997-10-16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>7791.0</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>OTTHUMG00000023822</td>\n",
       "      <td>uc003wcx.5</td>\n",
       "      <td>X95735</td>\n",
       "      <td>NM_003461</td>\n",
       "      <td>CCDS5883</td>\n",
       "      <td>Q15942</td>\n",
       "      <td>8917469|8940160</td>\n",
       "      <td>MGI:103072</td>\n",
       "      <td>RGD:620698</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>602002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:13200</td>\n",
       "      <td>ENST00000322764.10|NM_003461.5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43650</th>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>ZYXP1</td>\n",
       "      <td>zyxin pseudogene 1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>Approved</td>\n",
       "      <td>8q24.23</td>\n",
       "      <td>08q24.23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015-05-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015-05-07</td>\n",
       "      <td>106480342.0</td>\n",
       "      <td>ENSG00000274572</td>\n",
       "      <td>OTTHUMG00000187758</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NG_044600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PGOHUM00000303383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:51695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43651</th>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>zinc finger ZZ-type and EF-hand domain containing 1</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>17p13.2</td>\n",
       "      <td>17p13.2</td>\n",
       "      <td>KIAA0399|ZZZ4|FLJ10821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>zinc finger, ZZ-type with EF hand domain 1</td>\n",
       "      <td>Zinc fingers ZZ-type|EF-hand domain containing</td>\n",
       "      <td>91|863</td>\n",
       "      <td>2004-02-27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-02-12</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>23140.0</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>OTTHUMG00000090741</td>\n",
       "      <td>uc002fxe.4</td>\n",
       "      <td>BC035319</td>\n",
       "      <td>NM_015113</td>\n",
       "      <td>CCDS11043</td>\n",
       "      <td>O43149</td>\n",
       "      <td>9455477</td>\n",
       "      <td>MGI:2444286</td>\n",
       "      <td>RGD:1311189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>619459</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:29027</td>\n",
       "      <td>ENST00000381638.7|NM_015113.4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43652</th>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>zinc finger ZZ-type containing 3</td>\n",
       "      <td>protein-coding gene</td>\n",
       "      <td>gene with protein product</td>\n",
       "      <td>Approved</td>\n",
       "      <td>1p31.1</td>\n",
       "      <td>01p31.1</td>\n",
       "      <td>DKFZP564I052|ATAC1</td>\n",
       "      <td>ATAC component 1 homolog (Drosophila)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Zinc fingers ZZ-type|Myb/SANT domain containing|ATAC complex</td>\n",
       "      <td>91|532|1058</td>\n",
       "      <td>2004-01-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-02-12</td>\n",
       "      <td>2023-01-20</td>\n",
       "      <td>26009.0</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>OTTHUMG00000009652</td>\n",
       "      <td>uc001dhq.4</td>\n",
       "      <td>AL080063</td>\n",
       "      <td>NM_015534</td>\n",
       "      <td>CCDS677|CCDS76172</td>\n",
       "      <td>Q8IYH5</td>\n",
       "      <td>16428443|21304275</td>\n",
       "      <td>MGI:1920453</td>\n",
       "      <td>RGD:1565468</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>619892</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:24523</td>\n",
       "      <td>ENST00000370801.8|NM_015534.6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43653 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          hgnc_id    symbol   \n",
       "0          HGNC:5      A1BG  \\\n",
       "1      HGNC:37133  A1BG-AS1   \n",
       "2      HGNC:24086      A1CF   \n",
       "3          HGNC:7       A2M   \n",
       "4      HGNC:27057   A2M-AS1   \n",
       "...           ...       ...   \n",
       "43648  HGNC:25820    ZYG11B   \n",
       "43649  HGNC:13200       ZYX   \n",
       "43650  HGNC:51695     ZYXP1   \n",
       "43651  HGNC:29027     ZZEF1   \n",
       "43652  HGNC:24523      ZZZ3   \n",
       "\n",
       "                                                      name   \n",
       "0                                   alpha-1-B glycoprotein  \\\n",
       "1                                     A1BG antisense RNA 1   \n",
       "2                           APOBEC1 complementation factor   \n",
       "3                                    alpha-2-macroglobulin   \n",
       "4                                      A2M antisense RNA 1   \n",
       "...                                                    ...   \n",
       "43648         zyg-11 family member B, cell cycle regulator   \n",
       "43649                                                zyxin   \n",
       "43650                                   zyxin pseudogene 1   \n",
       "43651  zinc finger ZZ-type and EF-hand domain containing 1   \n",
       "43652                     zinc finger ZZ-type containing 3   \n",
       "\n",
       "               locus_group                 locus_type    status  location   \n",
       "0      protein-coding gene  gene with protein product  Approved  19q13.43  \\\n",
       "1           non-coding RNA       RNA, long non-coding  Approved  19q13.43   \n",
       "2      protein-coding gene  gene with protein product  Approved  10q11.23   \n",
       "3      protein-coding gene  gene with protein product  Approved  12p13.31   \n",
       "4           non-coding RNA       RNA, long non-coding  Approved  12p13.31   \n",
       "...                    ...                        ...       ...       ...   \n",
       "43648  protein-coding gene  gene with protein product  Approved    1p32.3   \n",
       "43649  protein-coding gene  gene with protein product  Approved      7q34   \n",
       "43650           pseudogene                 pseudogene  Approved   8q24.23   \n",
       "43651  protein-coding gene  gene with protein product  Approved   17p13.2   \n",
       "43652  protein-coding gene  gene with protein product  Approved    1p31.1   \n",
       "\n",
       "      location_sortable                   alias_symbol   \n",
       "0              19q13.43                            NaN  \\\n",
       "1              19q13.43                       FLJ23569   \n",
       "2              10q11.23  ACF|ASP|ACF64|ACF65|APOBEC1CF   \n",
       "3              12p13.31           FWP007|S863-7|CPAMD5   \n",
       "4              12p13.31                            NaN   \n",
       "...                 ...                            ...   \n",
       "43648           01p32.3                       FLJ13456   \n",
       "43649             07q34                            NaN   \n",
       "43650          08q24.23                            NaN   \n",
       "43651           17p13.2         KIAA0399|ZZZ4|FLJ10821   \n",
       "43652           01p31.1             DKFZP564I052|ATAC1   \n",
       "\n",
       "                                  alias_name                prev_symbol   \n",
       "0                                        NaN                        NaN  \\\n",
       "1                                        NaN  NCRNA00181|A1BGAS|A1BG-AS   \n",
       "2                                        NaN                        NaN   \n",
       "3                                        NaN                        NaN   \n",
       "4                                        NaN                        NaN   \n",
       "...                                      ...                        ...   \n",
       "43648                                    NaN                      ZYG11   \n",
       "43649                                    NaN                        NaN   \n",
       "43650                                    NaN                        NaN   \n",
       "43651                                    NaN                        NaN   \n",
       "43652  ATAC component 1 homolog (Drosophila)                        NaN   \n",
       "\n",
       "                                                                                                          prev_name   \n",
       "0                                                                                                               NaN  \\\n",
       "1      non-protein coding RNA 181|A1BG antisense RNA (non-protein coding)|A1BG antisense RNA 1 (non-protein coding)   \n",
       "2                                                                                                               NaN   \n",
       "3                                                                                                               NaN   \n",
       "4                                                      A2M antisense RNA 1 (non-protein coding)|A2M antisense RNA 1   \n",
       "...                                                                                                             ...   \n",
       "43648                                                     zyg-11 homolog (C. elegans)|zyg-11 homolog B (C. elegans)   \n",
       "43649                                                                                                           NaN   \n",
       "43650                                                                                                           NaN   \n",
       "43651                                                                    zinc finger, ZZ-type with EF hand domain 1   \n",
       "43652                                                                                                           NaN   \n",
       "\n",
       "                                                                       gene_group   \n",
       "0                                           Immunoglobulin like domain containing  \\\n",
       "1                                                                  Antisense RNAs   \n",
       "2                                                    RNA binding motif containing   \n",
       "3                                                    Alpha-2-macroglobulin family   \n",
       "4                                                                  Antisense RNAs   \n",
       "...                                                                           ...   \n",
       "43648  ZYG11 cell cycle regulator family|Armadillo like helical domain containing   \n",
       "43649                             Zyxin family|MicroRNA protein coding host genes   \n",
       "43650                                                                         NaN   \n",
       "43651                              Zinc fingers ZZ-type|EF-hand domain containing   \n",
       "43652                Zinc fingers ZZ-type|Myb/SANT domain containing|ATAC complex   \n",
       "\n",
       "      gene_group_id date_approved_reserved date_symbol_changed   \n",
       "0               594             1989-06-30                 NaN  \\\n",
       "1              1987             2009-07-20          2010-11-25   \n",
       "2               725             2007-11-23                 NaN   \n",
       "3              2148             1986-01-01                 NaN   \n",
       "4              1987             2012-06-23                 NaN   \n",
       "...             ...                    ...                 ...   \n",
       "43648        6|1492             2005-06-06          2005-07-11   \n",
       "43649     1402|1691             1997-10-16                 NaN   \n",
       "43650           NaN             2015-05-07                 NaN   \n",
       "43651        91|863             2004-02-27                 NaN   \n",
       "43652   91|532|1058             2004-01-13                 NaN   \n",
       "\n",
       "      date_name_changed date_modified    entrez_id  ensembl_gene_id   \n",
       "0                   NaN    2023-01-20          1.0  ENSG00000121410  \\\n",
       "1            2012-08-15    2013-06-27     503538.0  ENSG00000268895   \n",
       "2                   NaN    2023-01-20      29974.0  ENSG00000148584   \n",
       "3                   NaN    2023-01-20          2.0  ENSG00000175899   \n",
       "4            2018-03-21    2018-03-21     144571.0  ENSG00000245105   \n",
       "...                 ...           ...          ...              ...   \n",
       "43648        2012-12-10    2023-01-20      79699.0  ENSG00000162378   \n",
       "43649               NaN    2023-01-20       7791.0  ENSG00000159840   \n",
       "43650               NaN    2015-05-07  106480342.0  ENSG00000274572   \n",
       "43651        2016-02-12    2023-01-20      23140.0  ENSG00000074755   \n",
       "43652        2016-02-12    2023-01-20      26009.0  ENSG00000036549   \n",
       "\n",
       "                  vega_id     ucsc_id                     ena   \n",
       "0      OTTHUMG00000183507  uc002qsd.5                     NaN  \\\n",
       "1      OTTHUMG00000183508  uc002qse.3                BC040926   \n",
       "2      OTTHUMG00000018240  uc057tgv.1                AF271790   \n",
       "3      OTTHUMG00000150267  uc001qvk.2  BX647329|X68728|M11313   \n",
       "4      OTTHUMG00000168289  uc009zgj.2                     NaN   \n",
       "...                   ...         ...                     ...   \n",
       "43648  OTTHUMG00000008938  uc001cuj.4                AB051517   \n",
       "43649  OTTHUMG00000023822  uc003wcx.5                  X95735   \n",
       "43650  OTTHUMG00000187758         NaN                     NaN   \n",
       "43651  OTTHUMG00000090741  uc002fxe.4                BC035319   \n",
       "43652  OTTHUMG00000009652  uc001dhq.4                AL080063   \n",
       "\n",
       "      refseq_accession                               ccds_id uniprot_ids   \n",
       "0            NM_130786                             CCDS12976      P04217  \\\n",
       "1            NR_015380                                   NaN         NaN   \n",
       "2            NM_014576  CCDS7242|CCDS7241|CCDS73133|CCDS7243      Q9NQ94   \n",
       "3            NM_000014                             CCDS44827      P01023   \n",
       "4            NR_026971                                   NaN         NaN   \n",
       "...                ...                                   ...         ...   \n",
       "43648        NM_024646                             CCDS30717      Q9C0D3   \n",
       "43649        NM_003461                              CCDS5883      Q15942   \n",
       "43650        NG_044600                                   NaN         NaN   \n",
       "43651        NM_015113                             CCDS11043      O43149   \n",
       "43652        NM_015534                     CCDS677|CCDS76172      Q8IYH5   \n",
       "\n",
       "               pubmed_id       mgd_id       rgd_id   \n",
       "0                2591067  MGI:2152878    RGD:69417  \\\n",
       "1                    NaN          NaN          NaN   \n",
       "2      11815617|11072063  MGI:1917115   RGD:619834   \n",
       "3        2408344|9697696  MGI:2449119     RGD:2004   \n",
       "4                    NaN          NaN          NaN   \n",
       "...                  ...          ...          ...   \n",
       "43648           11214970  MGI:2685277  RGD:1307814   \n",
       "43649    8917469|8940160   MGI:103072   RGD:620698   \n",
       "43650                NaN          NaN          NaN   \n",
       "43651            9455477  MGI:2444286  RGD:1311189   \n",
       "43652  16428443|21304275  MGI:1920453  RGD:1565468   \n",
       "\n",
       "                                                               lsdb cosmic   \n",
       "0                                                               NaN    NaN  \\\n",
       "1                                                               NaN    NaN   \n",
       "2                                                               NaN    NaN   \n",
       "3      LRG_591|http://ftp.ebi.ac.uk/pub/databases/lrgex/LRG_591.xml    NaN   \n",
       "4                                                               NaN    NaN   \n",
       "...                                                             ...    ...   \n",
       "43648                                                           NaN    NaN   \n",
       "43649                                                           NaN    NaN   \n",
       "43650                                                           NaN    NaN   \n",
       "43651                                                           NaN    NaN   \n",
       "43652                                                           NaN    NaN   \n",
       "\n",
       "      omim_id mirbase  homeodb snornabase bioparadigms_slc  orphanet   \n",
       "0      138670     NaN      NaN        NaN              NaN       NaN  \\\n",
       "1         NaN     NaN      NaN        NaN              NaN       NaN   \n",
       "2      618199     NaN      NaN        NaN              NaN       NaN   \n",
       "3      103950     NaN      NaN        NaN              NaN       NaN   \n",
       "4         NaN     NaN      NaN        NaN              NaN       NaN   \n",
       "...       ...     ...      ...        ...              ...       ...   \n",
       "43648  618673     NaN      NaN        NaN              NaN       NaN   \n",
       "43649  602002     NaN      NaN        NaN              NaN       NaN   \n",
       "43650     NaN     NaN      NaN        NaN              NaN       NaN   \n",
       "43651  619459     NaN      NaN        NaN              NaN       NaN   \n",
       "43652  619892     NaN      NaN        NaN              NaN       NaN   \n",
       "\n",
       "          pseudogene.org horde_id   merops imgt iuphar  kznf_gene_catalog   \n",
       "0                    NaN      NaN  I43.950  NaN    NaN                NaN  \\\n",
       "1                    NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "2                    NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "3                    NaN      NaN  I39.001  NaN    NaN                NaN   \n",
       "4                    NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "...                  ...      ...      ...  ...    ...                ...   \n",
       "43648                NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "43649                NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "43650  PGOHUM00000303383      NaN      NaN  NaN    NaN                NaN   \n",
       "43651                NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "43652                NaN      NaN      NaN  NaN    NaN                NaN   \n",
       "\n",
       "       mamit-trnadb   cd lncrnadb enzyme_id  intermediate_filament_db   \n",
       "0               NaN  NaN      NaN       NaN                       NaN  \\\n",
       "1               NaN  NaN      NaN       NaN                       NaN   \n",
       "2               NaN  NaN      NaN       NaN                       NaN   \n",
       "3               NaN  NaN      NaN       NaN                       NaN   \n",
       "4               NaN  NaN      NaN       NaN                       NaN   \n",
       "...             ...  ...      ...       ...                       ...   \n",
       "43648           NaN  NaN      NaN       NaN                       NaN   \n",
       "43649           NaN  NaN      NaN       NaN                       NaN   \n",
       "43650           NaN  NaN      NaN       NaN                       NaN   \n",
       "43651           NaN  NaN      NaN       NaN                       NaN   \n",
       "43652           NaN  NaN      NaN       NaN                       NaN   \n",
       "\n",
       "       rna_central_ids lncipedia gtrnadb         agr   \n",
       "0                  NaN       NaN     NaN      HGNC:5  \\\n",
       "1                  NaN  A1BG-AS1     NaN  HGNC:37133   \n",
       "2                  NaN       NaN     NaN  HGNC:24086   \n",
       "3                  NaN       NaN     NaN      HGNC:7   \n",
       "4                  NaN   A2M-AS1     NaN  HGNC:27057   \n",
       "...                ...       ...     ...         ...   \n",
       "43648              NaN       NaN     NaN  HGNC:25820   \n",
       "43649              NaN       NaN     NaN  HGNC:13200   \n",
       "43650              NaN       NaN     NaN  HGNC:51695   \n",
       "43651              NaN       NaN     NaN  HGNC:29027   \n",
       "43652              NaN       NaN     NaN  HGNC:24523   \n",
       "\n",
       "                          mane_select       gencc  \n",
       "0       ENST00000263100.8|NM_130786.4         NaN  \n",
       "1                                 NaN         NaN  \n",
       "2       ENST00000373997.8|NM_014576.4         NaN  \n",
       "3      ENST00000318602.12|NM_000014.6      HGNC:7  \n",
       "4                                 NaN         NaN  \n",
       "...                               ...         ...  \n",
       "43648   ENST00000294353.7|NM_024646.3  HGNC:25820  \n",
       "43649  ENST00000322764.10|NM_003461.5         NaN  \n",
       "43650                             NaN         NaN  \n",
       "43651   ENST00000381638.7|NM_015113.4         NaN  \n",
       "43652   ENST00000370801.8|NM_015534.6         NaN  \n",
       "\n",
       "[43653 rows x 54 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc_master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge in HGNC IDs from hgnc_master_filtbygtex\n",
    "cols2include = ['symbol','hgnc_id','name','locus_group','locus_type','location']\n",
    "                                        #,'entrez_id','ensembl_gene_id','uniprot_ids']\n",
    "symbols_hgnc=pd.merge(left=gtx_genes,right=hgnc_master[cols2include],   # include more gene IDs/gene data here\n",
    "                           how='inner',\n",
    "                           on='symbol')#.dropna()#.drop_duplicates('Transcript ID')\n",
    "\n",
    "#symbols_hgnc.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now map COLUMN names (tissues) and UBERON codes. onto medgene_flat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "medgene_flat_ub = pd.merge(left=medgene_flat,right=SAMPID_2_TISSUE_unique,on='tissue')\n",
    "#medgene_flat_ub.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1573380, 11)\n"
     ]
    }
   ],
   "source": [
    "# Join flattened median gene dataset with the HGNC IDs (and other gene info)\n",
    "medgene_merge  = pd.merge(left=medgene_flat_ub,\n",
    "                     right=symbols_hgnc,\n",
    "                     how='inner',\n",
    "                     on='Transcript ID').dropna()\n",
    "\n",
    "print(medgene_merge.shape)\n",
    "#medgene_merge.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SKIP THS---DONT NEED CUIs---Merge in HGNC CUIs from the umls_genes (HGNC_IDs--HGNC CUI mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#medgene_merge_2  =  pd.merge(left=umls_genes,right=medgene_merge,on='hgnc_id')\n",
    "#assert medgene_merge.shape[0] == medgene_merge_2.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "medgene_merge_2 = medgene_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Merge in UBERON CUIs from the umls_genes (HGNC_IDs--HGNC CUI mapping)\n",
    "medgene_merge_2.rename(columns={'SMUBRID':'UBERON_code'},inplace=True)\n",
    "\n",
    "# Drop rows where UBERON code starts with 'EFO', which drops 70,738 rows\n",
    "#medgene_merge_2 = medgene_merge_2[~medgene_merge_2['UBERON_code'].str.contains('EFO')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Append UBERON to front of UBERON code unless it starts with EFO, then leave it alone\n",
    "medgene_merge_2['UBERON_CODEID'] = ['UBERON '+i if 'EFO_' not in i else i for i in medgene_merge_2['UBERON_code'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(medgene_merge_2[medgene_merge_2['UBERON_code'].str.startswith('UBERON EFO_')]) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We need to manually add the UMLS CUIs for some of these UBERON codes (again.) \n",
    "# As a reminder, they do not exist in UMLS as UBERON CUIs, which we searched for,\n",
    "# bc they were in UMLS (from NCIT) before UBERON was added. (meaning they didnt appear\n",
    "# in the search when I searched for UBERON Concepts, so I need to add them manually here)\n",
    "\n",
    "#extra_uberon_xref =  pd.DataFrame( [['0000458','C0227837'], # endocervix\n",
    "#                                   ['0001255','C0005682'],  # urinary bladder\n",
    "#                                   ['0003889','C0015560'],  # fallopian tube\n",
    "#                                   ['0012249','C0227829']], # ectocervix\n",
    "#                                     columns=['UBERON_code','UBERON_CUI'])\n",
    "\n",
    "#umls_uberon_xref_2 = umls_uberon_xref.append(extra_uberon_xref) # add these extra UBERON Code-CUI mappings to the rest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SKIP THIS---DONT NEED CUIs--- Merge in UBERON CUIs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#medgene_merge_3  =  pd.merge(left=medgene_merge_2,right=umls_uberon_xref_2,on='UBERON_code') \n",
    "#assert medgene_merge_2.shape[0]==medgene_merge_3.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "medgene_merge_3 = medgene_merge_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The last thing to do for the median_gene_tpm data set is to add unique code/Concept node identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "medgene_merge_3['tissue_no_space'] = [i.replace(' ','-') for i in medgene_merge_3['tissue']]\n",
    "medgene_merge_3['Transcript ID DASH'] =  medgene_merge_3['Transcript ID'].str.replace('.','-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turn the  'Median_TPM' into strings so we can include them in the unique hash ID \n",
    "median_tpm_strs = [str(i) for i in medgene_merge_3['Median_TPM']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# there are no columns with 100% unique values so we need to combine 3         FIX:  ' - ' +\n",
    "# columns to create unique strings that we can then hash\n",
    "medgene_merge_3['CODE_GTEX_Expression'] = medgene_merge_3['Transcript ID DASH'] + '-' + \\\n",
    "                                                    medgene_merge_3['tissue_no_space'] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "medgene_merge_3['CodeID_GTEX_Expression'] = ['GTEXEXP '+i for i in medgene_merge_3['CODE_GTEX_Expression']]\n",
    "\n",
    "# USE BASE-64 METHOD\n",
    "# create a 'GTEX EXP' CUI using a hash on that column.\n",
    "#medgene_merge_3['CUI_GTEX_Expression']  = ['KC' + str(int(hashlib.sha256(uid.encode('utf8')).hexdigest(),\n",
    "#                                                      base=16))[:CUI_LEN] for uid in medgene_merge_3['CODE_GTEX_Expression']]\n",
    "\n",
    "# Make CUIs\n",
    "#medgene_merge_3['CUI_GTEX_Expression'] = [i for i in CUIbase64(medgene_merge_3['CodeID_GTEX_Expression'])]\n",
    "\n",
    "# Check for collisions\n",
    "assert len(medgene_merge_3['CODE_GTEX_Expression'].unique()) == medgene_merge_3.shape[0]\n",
    "assert medgene_merge_3['CODE_GTEX_Expression'].unique().shape == medgene_merge_3['CodeID_GTEX_Expression'].unique().shape \n",
    "#assert  medgene_merge_3['CODE_GTEX_Expression'].unique().shape== medgene_merge_3['CUI_GTEX_Expression'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Select just the columns we need (CUIs and CodeIDs)\n",
    "#medgene_select = medgene_merge_3[['CUI_GTEX_Expression','CODE_GTEX_Expression','CodeID_GTEX_Expression','CUI_hgnc','Median_TPM','UBERON_CUI']]\n",
    "\n",
    "medgene_select = medgene_merge_3#[['CUI_GTEX_Expression','CODE_GTEX_Expression','CodeID_GTEX_Expression','CUI_hgnc','Median_TPM','UBERON_CUI']]\n",
    "#medgene_select.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### There are 2 CUI-CUI  mappings and 1 CUI-CodeID mapping we need for the GTEX EXP CUIs\n",
    "- CUI-CUIs\n",
    "    - GTEX_Expression CUI -- HGNC CUI\n",
    "    - GTEX_Expression CUI -- UBERON CUI\n",
    "- CUI-CODE\n",
    "    - GTEX_Expression CUI -- GTEX_Expression CodeID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CUIs (combine the Expression CUIs with the eQTL CUIs and save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GTEX_Ex_CUIs = pd.DataFrame(np.transpose([medgene_select['CUI_GTEX_Expression'].drop_duplicates().values]\n",
    "#                                          ),columns=['CUI:ID']) \n",
    "\n",
    "#CUIs_all_eqtl_2=CUIs_all_eqtl.to_frame().rename(columns={0:'CUI:ID'})#,inplace=True)\n",
    "\n",
    "# Combine eqtl CUIs with med expression CUIs\n",
    "#CUIs_all = CUIs_all_eqtl_2.append(GTEX_Ex_CUIs)\n",
    "\n",
    "#assert CUIs_all.shape  == CUIs_all.drop_duplicates().shape \n",
    "#CUIs_all.to_csv('/Users/stearb/desktop/R03_local/data/ingest_files/GTEx/CUIs_GTEx.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CUI-CUIs (and combine the Expression CUI-CUIs with the eQTL CUI-CUIs and save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "'''\n",
    "# GTEX_EXP --> HGNC \n",
    "GTEX_Ex_2_HGNC = medgene_select[['CUI_GTEX_Expression','CUI_hgnc']].rename(columns={'CUI_GTEX_Expression':':START_ID','CUI_hgnc':':END_ID'})\n",
    "GTEX_Ex_2_HGNC[':TYPE'] = 'median_expression_in_gene'\n",
    "GTEX_Ex_2_HGNC['SAB'] = 'GTEX_EXP__HGNC'\n",
    "\n",
    "# INVERSE, HGNC --> GTEX_EXP\n",
    "HGNC_2_GTEX_Ex = medgene_select[['CUI_hgnc','CUI_GTEX_Expression']].rename(columns={'CUI_GTEX_Expression':':END_ID','CUI_hgnc':':START_ID'})\n",
    "HGNC_2_GTEX_Ex[':TYPE'] = 'gene_has_median_expression'\n",
    "HGNC_2_GTEX_Ex['SAB'] = 'GTEX_EXP__HGNC'\n",
    "\n",
    "############################################\n",
    "############################################\n",
    "\n",
    "# GTEX_EXP --> UBERON\n",
    "GTEX_Ex_2_UBERON  = medgene_select[['CUI_GTEX_Expression','UBERON_CUI']].rename(columns={'CUI_GTEX_Expression':':START_ID','UBERON_CUI':':END_ID'})\n",
    "GTEX_Ex_2_UBERON[':TYPE'] = 'median_expression_in_tissue'\n",
    "GTEX_Ex_2_UBERON['SAB'] = 'GTEX_EXP__UBERON'\n",
    "\n",
    "\n",
    "# UBERON -->  GTEX_EXP\n",
    "UBERON_2_GTEX_Ex  = medgene_select[['UBERON_CUI','CUI_GTEX_Expression']].rename(columns={'CUI_GTEX_Expression':':END_ID','UBERON_CUI':':START_ID'})\n",
    "UBERON_2_GTEX_Ex[':TYPE'] = 'tissue_has_median_expression'\n",
    "UBERON_2_GTEX_Ex['SAB'] = 'GTEX_EXP__UBERON'\n",
    "\n",
    "\n",
    "# Combine CUI-CUIs\n",
    "GTEX_Ex_CUI_CUI = pd.concat([GTEX_Ex_2_HGNC,HGNC_2_GTEX_Ex,GTEX_Ex_2_UBERON,UBERON_2_GTEX_Ex])\n",
    "\n",
    "\n",
    "# Add in the eqtl CUI-CUIs\n",
    "#CUI_CUI_eqtls.rename(columns={'CUI_gtex':':START_ID','CUI':':END_ID'},inplace=True)\n",
    "\n",
    "# Combine with the eQTL CUI-CUIs\n",
    "CUI_CUIs_all = pd.concat([GTEX_Ex_CUI_CUI, CUI_CUI_eqtls]) \n",
    "\n",
    "# Check that there are no duplicates when we combine the 2 CUI-CUI dataframes\n",
    "assert CUI_CUIs_all.duplicated().sum() == 0\n",
    "\n",
    "#CUI_CUIs_all.to_csv('/Users/stearb/desktop/R03_local/data/ingest_files/GTEx/CUI-CUI_GTEx.csv',index=False)\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CODEs (and combine the Expression Codes with the eQTL Codes and save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GTEX_Ex_CODEs = medgene_select[['CodeID_GTEX_Expression','CODE_GTEX_Expression']].rename(columns={'CodeID_GTEX_Expression':'CodeID',\n",
    "#                                                                                   'CODE_GTEX_Expression':'CODE'})\n",
    "#GTEX_Ex_CODEs['SAB'] = 'GTEX_EXP'\n",
    "\n",
    "#GTEX_Ex_CODEs = GTEX_Ex_CODEs[['CodeID','SAB','CODE']] \n",
    "\n",
    "# Add in eqtl CODEs\n",
    "#CODEs_all = pd.concat([GTEX_Ex_CODEs,gtex_codes_eqtl])\n",
    "\n",
    "#assert CODEs_all.shape == CODEs_all.drop_duplicates().shape\n",
    "#assert CODEs_all.nunique()['CodeID'] == CODEs_all.nunique()['CODE']\n",
    "\n",
    "#CODEs_all.to_csv('/Users/stearb/desktop/R03_local/data/ingest_files/GTEx/CODEs_GTEx.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  CUI_CODEs (and combine the Expression CUI-Codes with the eQTL CUI-Codes and save)\n",
    "### fix duplicates issues\n",
    "There arent duplicates in GTEX_Expression CUI_CODE and there arent duplicates in GTEX eqtl CUI_CODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GTEX_Ex_CUI_CODE  = medgene_select[['CUI_GTEX_Expression',\n",
    "#                                    'CodeID_GTEX_Expression']].rename(columns={'CUI_GTEX_Expression':'CUI',\n",
    "#                                                                               'CodeID_GTEX_Expression':'CODE'})\n",
    "# Add in eqtl CUI-CODEs\n",
    "#GTEX_CUI_CODEs_all  = pd.concat([GTEX_Ex_CUI_CODE,GTEX_eqtl_CUI_CODEs])\n",
    "\n",
    "#assert GTEX_CUI_CODEs_all.shape == GTEX_CUI_CODEs_all.drop_duplicates().shape\n",
    "\n",
    "# No Overlap between GTEX eqtl CUIs and GTEX Expression CUIs\n",
    "#assert set() == set(GTEX_eqtl_CUI_CODEs['CUI']).intersection(set(GTEX_Ex_CUI_CODE['CUI']))\n",
    "\n",
    "#GTEX_CUI_CODEs_all.to_csv('/Users/stearb/desktop/R03_local/data/ingest_files/GTEx/CUI-CODEs_GTEx.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is actually no overlap between the SUI:IDs or the names/Terms between the EQTL SUIs (eqtl pvals) and the EXP SUIs (TPM values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#venn2([set(GTEX_EXP_SUIs['SUI:ID']),set(GTEX_EQTL_SUIs['SUI:ID'])])\n",
    "#venn2([set(GTEX_EXP_SUIs['name']),set(GTEX_EQTL_SUIs['name'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "#medgene_select['Median_TPM'].hist(bins=100)    \n",
    "# less than 2,000  = 1,837,599\n",
    "#  total = 1,839,188\n",
    "# zeros = 742,955\n",
    "#medgene_select[(medgene_select['Median_TPM'] > 50000)]['Median_TPM'].hist()\n",
    "#max(medgene_select[(medgene_select['Median_TPM'] != 0)]['Median_TPM'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Put TPM values into bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only 2 TPM values below 0.0007\n",
    "\n",
    "tpm_bins = list([0.0000000,7e-4,8e-4,9e-4]) + list(np.linspace(1e-3,9e-3,9)) + \\\n",
    "           list(np.round(np.linspace(1e-2,9e-2,9),2)) + list(np.round(np.linspace(.1,1,10),2)) + \\\n",
    "           list(np.linspace(2,100,99)) + list(np.arange(100,1100,100)[1:]) +  \\\n",
    "            list(np.arange(2000,11000,1000)) + list(np.arange(20000,110000,10000)) + [300000]\n",
    "\n",
    "# Seperate main df into df's where 'Median_TPM' is == 0 , and where 'Median_TPM' is != 0\n",
    "tpm_0s = medgene_select[medgene_select['Median_TPM'] == 0.00] \n",
    "tpm_intervals = medgene_select[medgene_select['Median_TPM'] != 0.00] \n",
    "\n",
    "# Create TPM Bins  \n",
    "tpm_intervals['tpm_bins'] = pd.cut(tpm_intervals['Median_TPM'], tpm_bins)\n",
    "\n",
    "# Create SUIs for intervals\n",
    "#tpm_intervals['SUI_tpm_bins']  = ['KS' + str(int(hashlib.sha256(str(uid).encode('utf8')).hexdigest(),base=16))[:SUI_LEN] for uid in tpm_intervals['tpm_bins']]\n",
    "\n",
    "# Save expression bins as strings first\n",
    "bin_strings_exp = tpm_intervals['tpm_bins'].astype(str)\n",
    "\n",
    "# Create SUIs the new way.\n",
    "#tpm_intervals['SUI_tpm_bins'] = [i for i in CUIbase64(bin_strings_exp)]\n",
    "\n",
    "\n",
    "# Seperate lower bounds and upper  bounds of intervals into 2 columns, unless its Nan (float), \n",
    "# then just set both bounds to 0.0\n",
    "# CHEECK WHAT THE FLOATS IN HERE  ARE, THEY SHOULDNT BE 0's\n",
    "tpm_intervals['bins_lowerbound'] = [i.left if type(i) is not float else 0.0 for i in tpm_intervals['tpm_bins'] ]\n",
    "tpm_intervals['bins_upperbound'] = [i.right if type(i) is not float else 0.0 for i in tpm_intervals['tpm_bins'] ]\n",
    "\n",
    "tpm_intervals.drop('tpm_bins',axis=1,inplace=True)\n",
    "\n",
    "# Create SUIs for  TPM == 0. It will be the same SUI (the 0.0 SUI) for all of these\n",
    "#tpm_0s['SUI_tpm_bins']  = ['KS' + str(int(hashlib.sha256(str(uid).encode('utf8')).hexdigest(),base=16))[:SUI_LEN] for uid in tpm_0s['Median_TPM']]\n",
    "\n",
    "bin_strings_exp_0 = tpm_0s['Median_TPM'].astype(str)\n",
    "\n",
    "#tpm_0s['SUI_tpm_bins'] = [i for i in CUIbase64(bin_strings_exp_0)]\n",
    "\n",
    "# Add columns to concat with main tpm_intervals\n",
    "tpm_0s['bins_upperbound'] = 0.0\n",
    "tpm_0s['bins_lowerbound'] = 0.0\n",
    "\n",
    "# Join them back together\n",
    "tpm_all_intervals =  pd.concat([tpm_intervals,tpm_0s]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create 'name' column for interval Terms.\n",
    "tpm_all_intervals['name'] = tpm_all_intervals['bins_lowerbound'].astype(str) +','+ tpm_all_intervals['bins_upperbound'].astype(str)\n",
    "\n",
    "tpm_all_intervals.drop(['bins_lowerbound','bins_upperbound'],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>Median_TPM</th>\n",
       "      <th>SMTS</th>\n",
       "      <th>UBERON_code</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>location</th>\n",
       "      <th>UBERON_CODEID</th>\n",
       "      <th>tissue_no_space</th>\n",
       "      <th>Transcript ID DASH</th>\n",
       "      <th>CODE_GTEX_Expression</th>\n",
       "      <th>CodeID_GTEX_Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>Testis</td>\n",
       "      <td>0.166403</td>\n",
       "      <td>Testis</td>\n",
       "      <td>0000473</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>0.1,0.2</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>1p36.33</td>\n",
       "      <td>UBERON 0000473</td>\n",
       "      <td>Testis</td>\n",
       "      <td>ENSG00000223972-5</td>\n",
       "      <td>ENSG00000223972-5-Testis</td>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Testis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>Adipose Subcutaneous</td>\n",
       "      <td>4.064030</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>0002190</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>4.0,5.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>1p36.33</td>\n",
       "      <td>UBERON 0002190</td>\n",
       "      <td>Adipose-Subcutaneous</td>\n",
       "      <td>ENSG00000227232-5</td>\n",
       "      <td>ENSG00000227232-5-Adipose-Subcutaneous</td>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>Adrenal Gland</td>\n",
       "      <td>2.685490</td>\n",
       "      <td>Adrenal Gland</td>\n",
       "      <td>0002369</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>2.0,3.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>1p36.33</td>\n",
       "      <td>UBERON 0002369</td>\n",
       "      <td>Adrenal-Gland</td>\n",
       "      <td>ENSG00000227232-5</td>\n",
       "      <td>ENSG00000227232-5-Adrenal-Gland</td>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adrenal-Gland</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Transcript ID                tissue  Median_TPM            SMTS   \n",
       "0  ENSG00000223972.5                Testis    0.166403          Testis  \\\n",
       "1  ENSG00000227232.5  Adipose Subcutaneous    4.064030  Adipose Tissue   \n",
       "2  ENSG00000227232.5         Adrenal Gland    2.685490   Adrenal Gland   \n",
       "\n",
       "  UBERON_code   symbol     hgnc_id     name locus_group  locus_type location   \n",
       "0     0000473  DDX11L1  HGNC:37102  0.1,0.2  pseudogene  pseudogene  1p36.33  \\\n",
       "1     0002190   WASH7P  HGNC:38034  4.0,5.0  pseudogene  pseudogene  1p36.33   \n",
       "2     0002369   WASH7P  HGNC:38034  2.0,3.0  pseudogene  pseudogene  1p36.33   \n",
       "\n",
       "    UBERON_CODEID       tissue_no_space Transcript ID DASH   \n",
       "0  UBERON 0000473                Testis  ENSG00000223972-5  \\\n",
       "1  UBERON 0002190  Adipose-Subcutaneous  ENSG00000227232-5   \n",
       "2  UBERON 0002369         Adrenal-Gland  ENSG00000227232-5   \n",
       "\n",
       "                     CODE_GTEX_Expression   \n",
       "0                ENSG00000223972-5-Testis  \\\n",
       "1  ENSG00000227232-5-Adipose-Subcutaneous   \n",
       "2         ENSG00000227232-5-Adrenal-Gland   \n",
       "\n",
       "                           CodeID_GTEX_Expression  \n",
       "0                GTEXEXP ENSG00000223972-5-Testis  \n",
       "1  GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous  \n",
       "2         GTEXEXP ENSG00000227232-5-Adrenal-Gland  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_all_intervals.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Expression Bin edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Testis</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.1,0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 4.0,5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adrenal-Gland</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 2.0,3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          subject       predicate   \n",
       "0                GTEXEXP ENSG00000223972-5-Testis  has_expression  \\\n",
       "1  GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous  has_expression   \n",
       "2         GTEXEXP ENSG00000227232-5-Adrenal-Gland  has_expression   \n",
       "\n",
       "            object  \n",
       "0  EXPBINS 0.1,0.2  \n",
       "1  EXPBINS 4.0,5.0  \n",
       "2  EXPBINS 2.0,3.0  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_all_intervals['bin_CodeID'] = ['EXPBINS ' + i for i in tpm_all_intervals['name']]\n",
    "\n",
    "edges_gtexEXP_bins = tpm_all_intervals[['CodeID_GTEX_Expression','bin_CodeID']]\n",
    "edges_gtexEXP_bins['predicate'] = 'has_expression'\n",
    "\n",
    "edges_gtexEXP_bins = edges_gtexEXP_bins[['CodeID_GTEX_Expression','predicate','bin_CodeID']]\n",
    "\n",
    "edges_gtexEXP_bins.columns = ['subject','predicate','object']\n",
    "edges_gtexEXP_bins.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_gtexEXP_bins['object'] = [i.replace(',','.') for i in edges_gtexEXP_bins['object']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject</th>\n",
       "      <th>predicate</th>\n",
       "      <th>object</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Testis</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.1.0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 4.0.5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Adrenal-Gland</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 2.0.3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Artery-Aorta</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 4.0.5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTEXEXP ENSG00000227232-5-Artery-Coronary</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 3.0.4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573915</th>\n",
       "      <td>GTEXEXP ENSG00000210195-2-Thyroid</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.0.0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573916</th>\n",
       "      <td>GTEXEXP ENSG00000210195-2-Uterus</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.0.0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573917</th>\n",
       "      <td>GTEXEXP ENSG00000210195-2-Vagina</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.0.0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573918</th>\n",
       "      <td>GTEXEXP ENSG00000210195-2-Whole-Blood</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.0.0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573919</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Whole-Blood</td>\n",
       "      <td>has_expression</td>\n",
       "      <td>EXPBINS 0.0.0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573920 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                subject       predicate   \n",
       "0                      GTEXEXP ENSG00000223972-5-Testis  has_expression  \\\n",
       "1        GTEXEXP ENSG00000227232-5-Adipose-Subcutaneous  has_expression   \n",
       "2               GTEXEXP ENSG00000227232-5-Adrenal-Gland  has_expression   \n",
       "3                GTEXEXP ENSG00000227232-5-Artery-Aorta  has_expression   \n",
       "4             GTEXEXP ENSG00000227232-5-Artery-Coronary  has_expression   \n",
       "...                                                 ...             ...   \n",
       "1573915               GTEXEXP ENSG00000210195-2-Thyroid  has_expression   \n",
       "1573916                GTEXEXP ENSG00000210195-2-Uterus  has_expression   \n",
       "1573917                GTEXEXP ENSG00000210195-2-Vagina  has_expression   \n",
       "1573918           GTEXEXP ENSG00000210195-2-Whole-Blood  has_expression   \n",
       "1573919           GTEXEXP ENSG00000210196-2-Whole-Blood  has_expression   \n",
       "\n",
       "                  object  \n",
       "0        EXPBINS 0.1.0.2  \n",
       "1        EXPBINS 4.0.5.0  \n",
       "2        EXPBINS 2.0.3.0  \n",
       "3        EXPBINS 4.0.5.0  \n",
       "4        EXPBINS 3.0.4.0  \n",
       "...                  ...  \n",
       "1573915  EXPBINS 0.0.0.0  \n",
       "1573916  EXPBINS 0.0.0.0  \n",
       "1573917  EXPBINS 0.0.0.0  \n",
       "1573918  EXPBINS 0.0.0.0  \n",
       "1573919  EXPBINS 0.0.0.0  \n",
       "\n",
       "[1573920 rows x 3 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges_gtexEXP_bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Save all SUIs (and combine the Expression SUIs with the eQTL SUIs and save)\\n(Combine GTEX EXPRESSION SUIs  with GTEX EQTL)  ; Columns: SUI:ID, name\\n        \\nAdd the 0 SUI and the interval SUIs to the main SUI file  \\nAdd just the 0 SUI to the SUIs_all (EQTL SUIs + EXPRESSION SUIs)  \\nWe need a seperate script to add the TPM interval SUIs bc they contain numerical properties  \\nBUT, add both the 0 SUIs and the interval SUIs to the CODE-SUIs_all, bc the relationships (CODE-SUIs) can be put in the ,UMLS master CODE-SUIs file.\\n# Add in all intervals'"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''Save all SUIs (and combine the Expression SUIs with the eQTL SUIs and save)\n",
    "(Combine GTEX EXPRESSION SUIs  with GTEX EQTL)  ; Columns: SUI:ID, name\n",
    "        \n",
    "Add the 0 SUI and the interval SUIs to the main SUI file  \n",
    "Add just the 0 SUI to the SUIs_all (EQTL SUIs + EXPRESSION SUIs)  \n",
    "We need a seperate script to add the TPM interval SUIs bc they contain numerical properties  \n",
    "BUT, add both the 0 SUIs and the interval SUIs to the CODE-SUIs_all, bc the relationships (CODE-SUIs) can be put in the ,UMLS master CODE-SUIs file.\n",
    "# Add in all intervals'''\n",
    "#GTEX_EXP_SUIs = tpm_all_intervals[['SUI_tpm_bins','name']].rename(columns={'SUI_tpm_bins':'SUI:ID'\n",
    "#                                                               }).drop_duplicates()\n",
    "\n",
    "# Combine GTEX_EXP_SUIs with GTEX EQTL SUIs\n",
    "#GTEX_SUIs = pd.concat([GTEX_EXP_SUIs,GTEX_EQTL_SUIs]).drop_duplicates()\n",
    "\n",
    "#assert len(GTEX_SUIs[GTEX_SUIs['SUI:ID'].duplicated()]) == 0\n",
    "\n",
    "#GTEX_SUIs.to_csv('/Users/stearb/desktop/R03_local/data/ingest_files/GTEx/SUIs_GTEx.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create edges files (GTEX_EXP--HGNC  &  GTEX_EXP---UBERON) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>tissue</th>\n",
       "      <th>Median_TPM</th>\n",
       "      <th>SMTS</th>\n",
       "      <th>UBERON_code</th>\n",
       "      <th>symbol</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>name</th>\n",
       "      <th>locus_group</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>location</th>\n",
       "      <th>UBERON_CODEID</th>\n",
       "      <th>tissue_no_space</th>\n",
       "      <th>Transcript ID DASH</th>\n",
       "      <th>CODE_GTEX_Expression</th>\n",
       "      <th>CodeID_GTEX_Expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>Adipose Subcutaneous</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>0002190</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 (pseudogene)</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>1p36.33</td>\n",
       "      <td>UBERON 0002190</td>\n",
       "      <td>Adipose-Subcutaneous</td>\n",
       "      <td>ENSG00000223972-5</td>\n",
       "      <td>ENSG00000223972-5-Adipose-Subcutaneous</td>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Adipose-Subcutaneous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Transcript ID                tissue  Median_TPM            SMTS   \n",
       "0  ENSG00000223972.5  Adipose Subcutaneous         0.0  Adipose Tissue  \\\n",
       "\n",
       "  UBERON_code   symbol     hgnc_id   \n",
       "0     0002190  DDX11L1  HGNC:37102  \\\n",
       "\n",
       "                                         name locus_group  locus_type   \n",
       "0  DEAD/H-box helicase 11 like 1 (pseudogene)  pseudogene  pseudogene  \\\n",
       "\n",
       "  location   UBERON_CODEID       tissue_no_space Transcript ID DASH   \n",
       "0  1p36.33  UBERON 0002190  Adipose-Subcutaneous  ENSG00000223972-5  \\\n",
       "\n",
       "                     CODE_GTEX_Expression   \n",
       "0  ENSG00000223972-5-Adipose-Subcutaneous  \\\n",
       "\n",
       "                           CodeID_GTEX_Expression  \n",
       "0  GTEXEXP ENSG00000223972-5-Adipose-Subcutaneous  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "medgene_select.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "medgene_select['hgnc_codeID'] = 'HGNC '+ medgene_select['hgnc_id']\n",
    "#medgene_select['uberon_CodeID'] = 'UBERON '+ medgene_select['UBERON_code']\n",
    "\n",
    "##################################\n",
    "######## WRONG RELATIONSHIPS ###\n",
    "medgene_select['gtex_hgnc_predicate'] =  'RO:0002206'  # expressed in   old rel 'median_expression_in_gene'\n",
    "medgene_select['gtex_uberon_predicate'] =  'RO:0002206'  # expressed in   old rel 'median_expression_in_gene' 'median_expression_in_tissue'\n",
    "\n",
    "edges_gtexExp_HGNC = medgene_select[['CodeID_GTEX_Expression','gtex_hgnc_predicate','hgnc_codeID']]\n",
    "edges_gtexExp_UB = medgene_select[['CodeID_GTEX_Expression','gtex_uberon_predicate','UBERON_CODEID']]\n",
    "\n",
    "edges_gtexExp_HGNC.columns = edges_gtexExp_UB.columns = ['subject','predicate','object']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_exp = pd.concat([edges_gtexEXP_bins,edges_gtexExp_HGNC,edges_gtexExp_UB])\n",
    "edges_exp = edges_exp.reset_index(drop=True)\n",
    "#edges_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_exp = edges_exp.drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create nodes file\n",
    "#### dont  need dbsnp  nodes saved here, theyre saved in the eqtl nodes file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Adipose-Subcutaneous</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Adrenal-Gland</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEXEXP ENSG00000223972-5-Artery-Aorta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          node_id  node_label\n",
       "0  GTEXEXP ENSG00000223972-5-Adipose-Subcutaneous         NaN\n",
       "1         GTEXEXP ENSG00000223972-5-Adrenal-Gland         NaN\n",
       "2          GTEXEXP ENSG00000223972-5-Artery-Aorta         NaN"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_GTEX_EXP = pd.DataFrame(medgene_select['CodeID_GTEX_Expression'] )\n",
    "nodes_GTEX_EXP['GTEX_EXP_terms'] = np.nan\n",
    "\n",
    "nodes_GTEX_EXP.columns = ['node_id','node_label']\n",
    "nodes_GTEX_EXP.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### and Expression_bin nodes file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_gtexEXP_bins = pd.DataFrame(edges_gtexEXP_bins['object'].drop_duplicates().reset_index(drop=True))\n",
    "nodes_gtexEXP_bins['node_label'] = np.nan\n",
    "nodes_gtexEXP_bins.columns = ['node_id','node_label']\n",
    "#nodes_gtexEXP_bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "## UBERON NODES\n",
    "nodes_exp_ub = pd.DataFrame(edges_gtexExp_UB['object'].drop_duplicates())\n",
    "nodes_exp_ub.columns = ['node_id']\n",
    "nodes_exp_ub['node_label'] = np.nan\n",
    "nodes_exp_ub = nodes_exp_ub.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EXPBINS 0.1.0.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXPBINS 4.0.5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EXPBINS 2.0.3.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EXPBINS 3.0.4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EXPBINS 5.0.6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573578</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Testis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573579</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Thyroid</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573580</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Uterus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573581</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Vagina</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1573582</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Whole-Blood</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573583 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       node_id  node_label\n",
       "0                              EXPBINS 0.1.0.2         NaN\n",
       "1                              EXPBINS 4.0.5.0         NaN\n",
       "2                              EXPBINS 2.0.3.0         NaN\n",
       "3                              EXPBINS 3.0.4.0         NaN\n",
       "4                              EXPBINS 5.0.6.0         NaN\n",
       "...                                        ...         ...\n",
       "1573578       GTEXEXP ENSG00000210196-2-Testis         NaN\n",
       "1573579      GTEXEXP ENSG00000210196-2-Thyroid         NaN\n",
       "1573580       GTEXEXP ENSG00000210196-2-Uterus         NaN\n",
       "1573581       GTEXEXP ENSG00000210196-2-Vagina         NaN\n",
       "1573582  GTEXEXP ENSG00000210196-2-Whole-Blood         NaN\n",
       "\n",
       "[1573583 rows x 2 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_exp = pd.concat([nodes_gtexEXP_bins,nodes_exp_ub,nodes_GTEX_EXP])\n",
    "nodes_exp = nodes_exp.reset_index(drop=True)\n",
    "nodes_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_exp = fill_missing_cols(nodes_exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_label</th>\n",
       "      <th>node_dbxrefs</th>\n",
       "      <th>node_namespace</th>\n",
       "      <th>node_definition</th>\n",
       "      <th>value</th>\n",
       "      <th>lowerbound</th>\n",
       "      <th>unit</th>\n",
       "      <th>upperbound</th>\n",
       "      <th>node_synonyms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EXPBINS 0.1.0.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXPBINS 4.0.5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EXPBINS 2.0.3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EXPBINS 3.0.4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EXPBINS 5.0.6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1574118</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Testis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1574119</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Thyroid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1574120</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Uterus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1574121</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Vagina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1574122</th>\n",
       "      <td>GTEXEXP ENSG00000210196-2-Whole-Blood</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1574123 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       node_id  node_label  node_dbxrefs   \n",
       "0                              EXPBINS 0.1.0.2         NaN           NaN  \\\n",
       "1                              EXPBINS 4.0.5.0         NaN           NaN   \n",
       "2                              EXPBINS 2.0.3.0         NaN           NaN   \n",
       "3                              EXPBINS 3.0.4.0         NaN           NaN   \n",
       "4                              EXPBINS 5.0.6.0         NaN           NaN   \n",
       "...                                        ...         ...           ...   \n",
       "1574118       GTEXEXP ENSG00000210196-2-Testis         NaN           NaN   \n",
       "1574119      GTEXEXP ENSG00000210196-2-Thyroid         NaN           NaN   \n",
       "1574120       GTEXEXP ENSG00000210196-2-Uterus         NaN           NaN   \n",
       "1574121       GTEXEXP ENSG00000210196-2-Vagina         NaN           NaN   \n",
       "1574122  GTEXEXP ENSG00000210196-2-Whole-Blood         NaN           NaN   \n",
       "\n",
       "         node_namespace  node_definition  value  lowerbound  unit  upperbound   \n",
       "0                   NaN              NaN    NaN         NaN   NaN         NaN  \\\n",
       "1                   NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "2                   NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "3                   NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "4                   NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "...                 ...              ...    ...         ...   ...         ...   \n",
       "1574118             NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "1574119             NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "1574120             NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "1574121             NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "1574122             NaN              NaN    NaN         NaN   NaN         NaN   \n",
       "\n",
       "         node_synonyms  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  \n",
       "...                ...  \n",
       "1574118            NaN  \n",
       "1574119            NaN  \n",
       "1574120            NaN  \n",
       "1574121            NaN  \n",
       "1574122            NaN  \n",
       "\n",
       "[1574123 rows x 10 columns]"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print([i for i in nodes_exp['node_id'] if 'EFO' in i])\n",
    "#nodes_exp = nodes_exp[~nodes_exp['node_id'].str.startswith('UBERON EFO_')].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_exp.to_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_exp/OWLNETS_node_metadata.txt',\n",
    "             sep='\\t',index=False)\n",
    "\n",
    "edges_exp.to_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_exp/OWLNETS_edgelist.txt',\n",
    "             sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_exp = pd.read_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_exp/OWLNETS_node_metadata.txt',sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_exp = pd.read_csv('/Users/stearb/Desktop/DESKTOP_TRANSFER/DataDistilleryFiles/gtex/gtex_exp/OWLNETS_edgelist.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#print([i for i in edges_exp['object'] if 'EFO' in i])\n",
    "\n",
    "\n",
    "edges_exp[edges_exp['object'].str.startswith('UBERON EFO_')]['object'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
