
# Petagraph Data Source Descriptions and Schema Reference

Each section contains the following information,

- Title
- Source of the data
- Description of how the data was preprocessed
- Schema figure
- Schema figure description
- Cypher query to produce the schema figure

## Genotype-Tissue Expression (GTEx) Portal
**Source**: We ingested two datasets from **[https://gtexportal.org/home/datasets](https://gtexportal.org/home/datasets):**
    - GTEx_Analysis_v8_eQTL (all files in this directory)
    - GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct

**Preproccessing**:
Code SABs: Gene (HGNC), Tissue (UBERON). The GTEx gene expression SABs are currently just [GTEx - Gene - Tissue - Expression] (SABs for GTEx gene expression names will be changed in the next version to just gene-tissue-GTEx_version)
    
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/gtex_exp.png" alt="drawing" width="800"/>
     
```cypher
// Cypher query to reproduce the schema figure
match (gtex_exp:Concept)-[:CODE]-(gtex_exp_code:Code {SAB:'GTEXEXP' })
match (gtex_exp)-[r1:expressed_in {SAB:'GTEXEXP'}]->(hgnc_concept:Concept)-[r2:CODE]-(hgnc_code:Code {SAB:'HGNC'}), (hgnc_concept)-[pt:PREF_TERM]-(t:Term)
match (gtex_exp)-[r3:expressed_in {SAB:"GTEXEXP"}]-(ub_concept:Concept)-[r4:CODE]-(ub_code:Code {SAB:'UBERON'}), (ub_concept)-[pt2:PREF_TERM]-(t2:Term)
match (gtex_exp)-[r5:has_expression]->(exp_concept:Concept)-[r6:CODE]-(exp_code:Code {SAB: 'EXPBINS' })
return * limit 1
```

--- 
## GTEXEQTL

**Source**:

**Preproccessing**:

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/gtex_eqtl.png" alt="drawing" width="800"/>


An UBERON Concept, Code and Term (top left), an HGNC Concept and preferred Term (top right) and GTEx eQTL Concept, Code and Terms (center). The GTEx Terms shown here represent a binned  p-value and variant ID for the eQTL.

```cypher
// Cypher query to reproduce the schema figure
match (gtex_eqtl:Concept)-[r0]-(gtex_eqtl_code:Code) where gtex_eqtl_code.SAB = 'GTEXEQTL' 
match (gtex_eqtl)-[r1]-(hgnc_concept:Concept)-[r2]-(hgnc_code:Code {CodeID:'HGNC:52402'}) where hgnc_code.SAB = 'HGNC'
match (gtex_eqtl)-[r3:located_in]-(ub_concept:Concept)-[r4]-(ub_code:Code) where ub_code.SAB = 'UBERON'
match (gtex_eqtl)-[r5:p_value]-(exp_concept:Concept)-[r6]-(exp_code:Code) where exp_code.SAB = 'PVALUEBINS'  
match  (gtex_eqtl)-[r7]-(hsclo_concept:Concept)-[r8]-(hsclo_code:Code) where hsclo_code.SAB = 'HSCLO'  
return * limit 1
```

---
## GTEXCOEXP
**Source**:
- Co-expression of genes were computed using Pearson’s correlation > 0.9 in 37 human tissues according to the GTEx expression data:
- GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct
- Relationship Name: **coexpressed_with**
- Tissue where the co-expression is detected is in the SAB of the relationship  “coexpressed_with”

**Preproccessing**:

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/gtexcoexp.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code)-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
where type(r1) starts with 'coexpressed_with'
return * limit 1
```


---
## Human-Mouse Orthologs (HGNCHCOP)

**Source**:
- Orthologs from HGNC Comparisons of Orthology Predictions (**HCOP**) [https://www.genenames.org/tools/hcop/](https://www.genenames.org/tools/hcop/) (scroll to the bottom, under Bulk Downloads. Select Human - Mouse ortholog data)

**Preproccessing**:

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/HGNCHCOP.png" alt="drawing" width="800"/>

HGNC Concept (blue), Code (purple) and Term (green) from HGNC on the left and its corresponding Mouse gene Concept and code on the right  


```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HCOP'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
return * limit 1
```

---
## Human gene-phenotype (HGNCHPO)

**Source**:
We use the Human Phenotype (HP) Ontology mappings for gene-to-phenotype and phenotype-to-genes. The HP annotations can be found here: [https://hpo.jax.org/app/data/annotations](https://hpo.jax.org/app/data/annotations). 

These data are generated by the HP group to use OMIM disease-gene associations to map all HP phenotypes to genes with those phenotypes associated with diseases.  Therefore, a gene can be associated with several phenotypes, and a phenotype can be associated with several genes.

- OMIM and Orphanet are combined together in the HPO database. 

**Preproccessing**:

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/HGNCHPO" alt="drawing" width="800"/>

A Concept (blue), Code (purple) and Term (green) node from HPO (left side) and HGNC (right side) and the bidirectional relationships between the two Concept nodes.

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HGNC'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HPO'})
return * limit 1
```

---
## Mouse gene-phenotype (HCOPMP)

**Source**:
- We ingested two datasets from the International Mouse Phenotype Consortium (**IMPC**) which contains data from the Knockout Mouse Phenotyping Program (**KOMP2**) at [http://ftp.ebi.ac.uk/pub/databases/impc/all-data-releases/latest/results/](http://ftp.ebi.ac.uk/pub/databases/impc/all-data-releases/latest/results/)
    - genotype-phenotype-assertions-ALL.csv
    - statistical-results-ALL.csv
- And three datasets from Mouse Genome Informatics (**MGI**) at [http://www.informatics.jax.org/downloads/reports/index.html#pheno](http://www.informatics.jax.org/downloads/reports/index.html#pheno)
    - MGI_PhenoGenoMP.rpt
    - MGI_GenePheno.rpt
    - MGI_Geno_DiseaseDO.rpt

**Preproccessing**:
Mouse genotype-to-phenotype (HCOPMP) data were obtained in January 2021 from multiple datasets from two separate databases. The first set of datasets were obtained from the international mouse phenotyping consortium (IMPC), which includes data from KOMP2, and can be found at http://ftp.ebi.ac.uk/pub/databases/impc/all-data-releases/latest/results/. We used the genotype-phenotype-assertions-ALL.csv.gz and the statistical-results-ALL.csv.gz datasets from this database. Both datasets contain, among other data, phenotype to gene mappings in the mouse. The second set of datasets were obtained from the mouse genome informatics (MGI) database and can be found at http://www.informatics.jax.org/downloads/reports/index.html#pheno. We used the MGI_PhenoGenoMP.rpt (Table 5),  MGI_GenePheno.rpt (Table 9) and MGI_Geno_DiseaseDO.rpt (Table 10) datasets. All 3 datasets contain, among other data, phenotype-to-gene mappings. The datasets from IMPC and MGI were combined to create a master genotype-to-phenotype dataset. This master dataset contains 10,380 MP terms that are mapped to at least one gene and 17,936 genes that are mapped to at least one MP term.
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/HCOPMP.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HCOP'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'MP'})
return * limit 1
```

---
## Human Phenotype Ontology (HPO) - Mouse Phenotype (MP) mappings (HPOMP)
**Source**:  The PheKnowLator tool,  [https://github.com/callahantiff/PheKnowLator](https://github.com/callahantiff/PheKnowLator) was used to map HPO terms to MP terms using semantic matching.


**Preproccessing**:  Matches were then manually curated.
  
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/HPOMP.png" alt="drawing" width="800"/>

A Concept (blue), Code (purple) and Term (green) from HPO on the left and its corresponding MP Concept, Code and Term on the right.

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HPO'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'MP'})
return * limit 1
```

---
## Human-Rat ENSEMBL orthology (RATHCOP)
**Source**:

**Preproccessing**:

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/RATHCOP.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'ENSEMBL'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'ENSEMBL'})
where a.CODE contains 'ENSR'
return * limit 1
```

---
## GENCODE-HSCLO mappings (GENCODEHSCLO)
**Source**:
**Preproccessing**:
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/GENCODEHSCLO.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'GENCODE'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HSCLO'})
return * limit 1
```

---
### LINCS L1000 Gene-Perturbagen Associations (LINCS)
**Source**:
**Preproccessing**:
- LINCS L1000 Connectivity Map dataset was obtained from the Ma’ayan Lab Harmonizome portal:
- [https://maayanlab.cloud/Harmonizome/search?t=all&q=l1000](https://maayanlab.cloud/Harmonizome/search?t=all&q=l1000)
- Relationship SAB: **LINCS L1000**
- Relationship Name: **positively_correlated_with_chemical_or_drug, positively_correlated_with_gene, negatively_correlated_with_chemical_or_drug, negatively_correlated_with_gene**
    
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/LINCS.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'CHEBI'})-[r0:CODE]-(b:Concept)-[r1 {SAB:'LINCS'}]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
return * limit 1
```

---
## Connectivity Map (CMAP)
**Source**:
- Signature perturbations of gene expression profiles as induced by chemical (small molecule) were obtained from the Ma’ayan Lab Harmonizome portal:
- [https://maayanlab.cloud/Harmonizome/dataset/CMAP+Signatures+of+Differentially+Expressed+Genes+for+Small+Molecules](https://maayanlab.cloud/Harmonizome/dataset/CMAP+Signatures+of+Differentially+Expressed+Genes+for+Small+Molecules)
- Relationship SAB: **CMAP**
- Relationship Name: **positively_correlated_with_chemical_or_drug, positively_correlated_with_gene, negatively_correlated_with_chemical_or_drug, negatively_correlated_with_gene**
  
**Preproccessing**:  
  
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/CMAP.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'CHEBI'})-[r0:CODE]-(b:Concept)-[r1 {SAB:'CMAP'}]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
return * limit 1
```

---
## Homo Sapiens Chromosomal Ontology (HSCLO)
**Source**: The HSCLO was created by Taha Ahooyi Mohseni.

**Preproccessing**: 
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/HSCLO.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HSCLO'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code)
MATCH (b)-[r3]-(c:Concept)-[:CODE]-(e:Code {SAB:'4DNL'})
MATCH  (b)-[r4]-(c2:Concept)-[:CODE]-(f:Code {SAB:'GTEXEQTL'})
MATCH  (b)-[r5]-(c3:Concept)-[:CODE]-(g:Code {SAB:'ENSEMBL'})
return * limit 1
```

---
## Molecular Signatures Database (MSIGDB) 
**Sourcee**: 
- MSigDB v7.4 datasets C1, C2, C3, C8 and H were obtained and ingested from the MSigDB molecular signature database:
- [https://www.gsea-msigdb.org/gsea/msigdb/](https://www.gsea-msigdb.org/gsea/msigdb/)
- Relationship SAB: **MSigDB C1, MSigDB C2, MSigDB C3, MSigDB C8, MSigDB H**
- Code SAB: **MSigDB_Systematic_Name**
- 
**Preproccessing**:
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/MSIGDB.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'HGNC'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'MSIGDB'})
return * limit 1
```


---
## CLINVAR
**Source**:
- ClinVar human genetic variant-disease associations were obtained from: [https://www.ncbi.nlm.nih.gov/clinvar/](https://www.ncbi.nlm.nih.gov/clinvar/)
- Only associations with Pathogenic and/or Likely Pathogenic consequence which met assertion criteria were included in the graph.
- Future versions of the KG will have the rest of the consequence levels.
- Relationship SAB: **ClinVar**
- Relationship Name: **gene_associated_with_disease_or_phenotype, disease_or_phenotype_associated_with_gene**
**Preproccessing**:
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/CLINVAR.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'UBERON'})-[r0:CODE]-(b:Concept)-[r1 {SAB:'CLINVAR'}]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
return * limit 1
```

---
## Azimuth (AZ)
**Source**:  ...
**Preproccessing**: ...

<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/AZ.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'AZ'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
return * limit 1
```

---
## Search Tool for the Retrieval of Interacting Genes/Proteins (STRING)
**Source**: 
**Preproccessing**: human only
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/STRING.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'UNIPROTKB'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'UNIPROTKB'})
return * limit 1
```

---
## Single Cell Heart expression data (ASP2019)
**Source**: 
**Preproccessing**: 
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/SCHEART.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'ASP2019'})-[r0:CODE]-(b:Concept)-[r1:expressed_in]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'HGNC'})
match (b)-[:expressed_in]-(e:Concept)-[:CODE]-(f:Code {SAB:'ASP2019CLUSTER'})
match (b)-[:log2FC]-(g:Concept)-[:CODE]-(h:Code {SAB:'LOG2FCBINS'})
return * limit 1
```

---
## GlyGen: Computational and Informatics Resources for Glycoscience (GLYGEN)
**Source**: 
**Preproccessing**: 
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/GLYGEN.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'GLY.TYPE.SITE'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'GLYTOUCAN'})
match (b)-[r3]-(e:Concept)-[:CODE]-(f:Code {SAB:'UNIPROTKB.ISOFORM'})
return * limit 1
```

---
## Kids First (KF)
**Source**: 
**Preproccessing**: 
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/kf.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (c0:Code {SAB:'KFCOHORT'})-[r0:CODE]-(cui1:Concept)-[r1:belongs_to_cohort]-(cui2:Concept )-[r2:CODE]-(c1:Code {SAB:'KFPT'})
match (cui2)-[r3:has_phenotype]-(cui3:Concept)-[r4:CODE]-(c2:Code {SAB:'HPO'})
match (cui1)-[r5:belongs_to_cohort]-(cui4:Concept)-[r6:CODE]-(c3:Code {SAB:'KFGENEBIN'})
match (cui4)-[r7:gene_has_variants]-(cui5:Concept)-[r8:CODE]-(c4:Code {SAB:'HGNC'})
return * LIMIT 1
```

---
## 4D Nucleome Program (4DN)
**Source**: 
**Preproccessing**: 
<img src="https://github.com/TaylorResearchLab/Petagraph/blob/main/figures/publication_figures/schema_figures/4DN.png" alt="drawing" width="800"/>

```cypher
// Cypher query to reproduce the schema figure
match (a:Code {SAB:'4DNF'})-[r0:CODE]-(b:Concept)-[r1]-(c:Concept)-[r2:CODE]-(d:Code {SAB:'4DNL'})
match (b)-[r3]-(e:Concept)-[:CODE]-(f:Code {SAB:'4DND'})
match (c)-[:loop_ds_end]-(g:Concept)-[:CODE]-(h:Code {SAB:'HSCLO'})
match (c)-[r4]-(i:Concept)-[:CODE]-(j:Code {SAB:'4DNQ'})
//match (j)-[r5]-(k:Concept)-[:CODE]-(l:Code {SAB:'UBERON'})
return * limit 1
```

